# R session information {.tabset}

```{r session_info}
if (is.list(params)) {
    markdownHeader("YAML params", level = 2)
    print(params)
}

markdownHeader("`_setup.R`", level = 2)
readr::read_lines("_setup.R") %>% writeLines()

markdownHeader("`sessioninfo::session_info()`", level = 2)
sessioninfo::session_info()

markdownHeader("`utils::sessionInfo()`", level = 2)
utils::sessionInfo()
```

[bcbio]: https://bcbio-nextgen.readthedocs.io/
[bcbioSingleCell]: https://bioinformatics.sph.harvard.edu/bcbioSingleCell/
[bcl2fastq]: https://support.illumina.com/sequencing/sequencing_software/bcl2fastq-conversion-software.html
[conda]: https://conda.io/
[DESeq2]: https://bioconductor.org/packages/DESeq2/
[edgeR]: https://bioconductor.org/packages/edgeR/
[Ensembl]: https://ensembl.org/
[inDrops]: http://1cell-bio.com/
[Matrix]: https://cran.r-project.org/package=Matrix
[Python]: https://python.org/
[R]: https://www.r-project.org/
[rapmap]: https://github.com/COMBINE-lab/RapMap/
[scater]: https://bioconductor.org/packages/scater/
[scran]: https://bioconductor.org/packages/scran/
[scRNA-seq]: http://bcbio-nextgen.readthedocs.io/en/latest/contents/configuration.html#single-cell-rna-sequencing
[Seurat]: http://satijalab.org/seurat/
[UMAP]: https://github.com/lmcinnes/umap/
[umis]: https://github.com/vals/umis/
[zinbwave]: https://bioconductor.org/packages/zinbwave/
[zingeR]: https://github.com/statOmics/zingeR/
