---
title: "Quality Control"
author: "`r getOption('author')`"
date: "`r Sys.Date()`"
bibliography: bibliography.bib
params:
    bcbFile: "data/bcb.rda"
    minUMIs: 500
    minGenes: 500
    maxGenes: 5000
    maxMitoRatio: 0.1
    minNovelty: 0.8
    minCellsPerGene: 3
    outputDir: "."
---

```{r setup, cache=FALSE, message=FALSE, warning=FALSE}
library(bcbioSingleCell)

# Shared RMarkdown settings
prepareSingleCellTemplate()
if (file.exists("setup.R")) {
    source("setup.R")
}

# Directory paths
dataDir <- file.path(params$outputDir, "data")

# Load bcbioSingleCell object
bcbName <- load(params$bcbFile)
bcb <- get(bcbName, inherits = FALSE)

# Back up all data inside bcbio object
flatFiles <- flatFiles(bcb)
saveData(flatFiles, dir = dataDir)
```

```{r header, child="_header.Rmd", eval=file.exists("_header.Rmd")}
```



```{r sample_metadata}
sampleMetadata(bcb)
```

[bcbio][] run data was imported from **`r metadata(bcb)[["uploadDir"]]`**.



# Count alignment

We aligned the counts using the latest transcriptome FASTA from [Ensembl][].



# Quality control metrics {.tabset}

## Reads per cell

These are counts of how many reads are assigned to a given cellular barcode. The purpose of these histograms is that we should see one single large peak that represents cells that were encapsulated. If we see a strong shoulder, or a bimodal distribution of the cells, that can indicate a couple problems. It might be that there is free floating RNA, which happens when cells are dying. It could also be that there are a set of cells that failed for some reason. Finally, it could also be that there are biologically different types of cells, and one type is much smaller than the other. If this is the case we would expect to see less RNA being sequenced from the smaller cells.

```{r plot_reads_per_cell, fig.height=12, fig.width=8}
plotReadsPerCell(bcb)
```


## Cell counts

```{r plot_cell_counts}
plotCellCounts(bcb)
```


## UMI counts per cell

```{r plot_umis_per_cell, fig.height=12, fig.width=8}
plotUMIsPerCell(bcb, min = params$minUMIs)
```


## Genes detected per cell

Here by "detected", we mean genes with a non-zero count measurement per cell. Seeing gene detection in the range of `500`-`5000` is normal for [inDrop][] analysis.

```{r plot_genes_per_cell, fig.height=12, fig.width=8}
plotGenesPerCell(bcb, min = params$minGenes, max = params$maxGenes)
```


## UMIs vs. genes detected

If we graph out the total number of UMI counts per cell vs. the genes detected per cell, we can assess whether there is a large population of low quality cells with low counts and/or gene detection.

```{r plot_umis_vs_genes}
plotUMIsVsGenes(bcb)
```


## Mitochondrial counts ratio

We evaluate overall mitochondrial gene expression as a biomarker of cellular stress during sample preparation.

```{r plot_mito_ratio, fig.height=16, fig.width=8}
plotMitoRatio(bcb, max = params$maxMitoRatio)
```


## Novelty

Another way to QC the data is to look for less novelty, that is cells that have less genes detected per count than other cells. We can see the samples where we sequenced each cell less have a higher overall novelty, that is because we have not started saturated the sequencing for any given gene for these samples. Outlier cells in these samples might be cells that we have a less complex RNA species than other cells. Sometimes we can detect contamination with low complexity cell types like red blood cells via this metric.

```{r plot_novelty, fig.height=12, fig.width=8}
plotNovelty(bcb, min = params$minNovelty)
```



# Filter cells

```{r filter_cells, fig.height=12, fig.width=8, results="asis"}
bcb <- filterCells(
    bcb,
    minUMIs = params$minUMIs,
    minGenes = params$minGenes,
    maxGenes = params$maxGenes,
    maxMitoRatio = params$maxMitoRatio,
    minNovelty = params$minNovelty,
    minCellsPerGene = params$minCellsPerGene)
assignAndSaveData(
    name = bcbName,
    object = bcb,
    dir = dataDir)
```



```{r footer, child="_footer.Rmd", eval=file.exists("_footer.Rmd")}
```
