<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>All news • bcbioSingleCell</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  <link href="../extra.css" rel="stylesheet">
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bcbioSingleCell</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioSingleCell">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">

  <div class="col-md-9">
    <div class="page-header">
      <h1>Change log <small>All releases</small></h1>
    </div>

    <div class="contents">
    <div id="bcbiosinglecell-0-0-22-2017-11-17" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-22-2017-11-17" class="anchor"></a>bcbioSingleCell 0.0.22 (2017-11-17)</h1>
<ul>
<li>Now internally stashing a <code>cell2sample</code> data.frame, which helps speed up operations on cellular barcode metrics calculations for quality control plots.</li>
<li>Improved support for optional <code>annotable</code>, <code>ensemblVersion</code>, <code>gtfFile</code>, and <code>sampleMetadataFile</code> arguments in <code><a href="../reference/loadSingleCell.html">loadSingleCell()</a></code> function.</li>
<li>Simplified some of the messages shown during sample loading, in an attempt to make them clearer and more informative. Also improved messages shown to the user during a <code><a href="../reference/filterCells.html">filterCells()</a></code> function call.</li>
<li>The <code><a href="../reference/metrics.html">metrics()</a></code> function will now look for a stashed <code>cell2sample</code> data.frame, which speeds up operations for quality control plots.</li>
<li>Improved handling of sample metadata columns as factors. In particular, levels should be correctly updated using <code>droplevels</code> in a <code><a href="../reference/selectSamples.html">selectSamples()</a></code> call. The <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> package has also been updated to work in a similar fashion, where all columns in the sample metadata data.frame are now defined as factors.</li>
<li>Simplified <code>bcbioSingleCell</code> to <code>seurat</code> object coercion to stash all of the bcbio metadata, and simply return the basic <code>seurat</code> object, rather than trying to also perform normalization and scaling. These steps have instead been added back to the Seurat R Markdown clustering template.</li>
<li>Updated cell cycle and cell type markers from our master copy on Google Sheets.</li>
<li>Added a troubleshooting section to the GitHub README, with a note on maximum DLLs.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-21-2017-11-08" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-21-2017-11-08" class="anchor"></a>bcbioSingleCell 0.0.21 (2017-11-08)</h1>
<ul>
<li>Updated package imports to match Bioconductor 3.6.</li>
<li>Initial support for <code><a href="../reference/plotCellTypesPerCluster.html">plotCellTypesPerCluster()</a></code>.</li>
<li>Initial support for <code><a href="../reference/plotMitoVsCoding.html">plotMitoVsCoding()</a></code>. I broke this code out from <code><a href="../reference/plotMitoRatio.html">plotMitoRatio()</a></code>. We could opt to keep this in <code>plotMitoRatio</code> with a <code>geom = "scatterplot"</code> argument.</li>
<li>Initial commit of <code>.applyFilterCutoffs()</code> internal function, used to subset the object to contain only cells and genes that have passed quailty control filtering.</li>
<li>Improved Seurat <code>FindAllMarkers()</code> sanitization.</li>
<li>Made the quality control plots more modular. Now they support multiple geoms, including <code>boxplot</code>, <code>histogram</code>, <code>ridgeline</code>, and <code>violin</code> (default). Median labels are applied with the internal <code>.medianLabels()</code> function.</li>
<li>Updated error message for CellRanger directory structure.</li>
<li>Cellular barcode columns are no longer split with an underscore. Instead, they are kept as a single ACGT string. We’re now generating <code>cellID</code> to <code>sampleID</code> matching with a different method. In the future, we’ll stash a cell2sample data.frame inside the object, that makes this operation faster than the current <code>mclapply()</code> code.</li>
<li>Updated annotable support in <code><a href="../reference/loadSingleCell.html">loadSingleCell()</a></code> and <code><a href="../reference/loadCellRanger.html">loadCellRanger()</a></code>.</li>
<li>Added support for handling both gene- and transcript-level counts. The updated release of the bcbio single-cell pipeline now outputs at gene level.</li>
<li>Initial quality control plot support for seurat objects.</li>
<li>Added support for return of only filtered cells and genes with the <code><a href="../reference/counts.html">counts(filterCells = TRUE)</a></code> function.</li>
<li>Added assignment support for <code>interestingGroups&lt;-</code>.</li>
<li>Initial support for sample metadata generation from seurat object.</li>
<li>Improved internal code for <code><a href="../reference/selectSamples.html">selectSamples()</a></code>.</li>
<li>Updated <code><a href="../reference/topMarkers.html">topMarkers()</a></code> to match Seurat v2.1 update.</li>
<li>Improved <code>bcbioSingleCell</code> to <code>seurat</code> coercion method with <code>setAs()</code>.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-20-2017-10-24" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-20-2017-10-24" class="anchor"></a>bcbioSingleCell 0.0.20 (2017-10-24)</h1>
<ul>
<li>Upgraded to basejump 0.1.0 and Seurat 2.1.0 dependencies.</li>
<li>Improved documentation of NAMESPACE imports per function.</li>
<li>Switched to base grep functions where applicable (<code>grepl()</code>, <code>gsub()</code>).</li>
<li>Use GTF in package documentation rather than GFF. Applies primarily to the <code><a href="../reference/loadSingleCell.html">loadSingleCell()</a></code> import function.</li>
<li>Restrict class support in S4 methods to <code>bcbioSingleCell</code>. Legacy <code>bcbioSCDataSet</code> class can be upgraded to <code>bcbioSingleCell</code> class using <code>as(bcb, "bcbioSingleCell")</code> coercion.</li>
<li>Use filtered cell output for <code><a href="../reference/metrics.html">metrics()</a></code> and quality control functions by default.</li>
<li>Updated the quality control R Markdown to include <code>filterCells = FALSE</code> where applicable.</li>
<li>Draft support for aggregated technical replicates in quality control functions using <code>sampleNameAggregate</code> column in sample metadata. This doesn’t change the actual counts values. It only applies to visualization in the quality control plots currently.</li>
<li>Miscellaneous R Markdown template updates. Primarily improvements to the setup chunk object loading workflow.</li>
<li>Removed lintr checks from testthat. This is breaking <code><a href="http://www.rdocumentation.org/packages/devtools/topics/test">devtools::test()</a></code>.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-19-2017-10-12" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-19-2017-10-12" class="anchor"></a>bcbioSingleCell 0.0.19 (2017-10-12)</h1>
<ul>
<li>Renamed main object class from <code>bcbioSCDataSet</code> to <code>bcbioSingleCell</code>.</li>
<li>Cell filtering with <code><a href="../reference/filterCells.html">filterCells()</a></code> will now slot a named logical vector into <code>metadata(object)[["filteredCells"]]</code>, which will be used to dynamically subset the slotted internal <code>SummarizedExperiment</code> data. Now that we’re using this approach, we can return a modified <code>bcbioSingleCell</code> object rather than defining a separate <code>bcbioSCFiltered</code> class.</li>
<li>Renamed <code><a href="../reference/deprecated.html">loadSingleCellRun()</a></code> to <code><a href="../reference/loadSingleCell.html">loadSingleCell()</a></code>, to match bcbioRNASeq package.</li>
<li>Now allowing implicit integers in our function code.</li>
<li>Added support for plotting technical replicates. This is handled by <code>sampleNameAggregate</code> in the sample metadata.</li>
<li>Now using ridgeline plots in place of histograms where applicable.</li>
<li>Travis CI checks take too long when loading SummarizedExperiment. Hopefully this will be fixed in the 3.6 release later this month.</li>
<li>New internal dark theme (<code><a href="../reference/darkTheme.html">darkTheme()</a></code>), based on the Seurat theme.</li>
<li>Initial commit of <code><a href="../reference/plotDot.html">plotDot()</a></code> function, based on <code><a href="http://www.rdocumentation.org/packages/Seurat/topics/DotPlot">Seurat::DotPlot()</a></code>.</li>
<li>Added new tSNE plots that allow for consistent cluster labeling.</li>
<li>Providing legacy support for <code>bcbioSCDataSet</code> and <code>bcbioSCFiltered</code>, which will be deprecated in a future release.</li>
<li>Offloaded some internal code to basejump, for improved consistency with bcbioRNASeq package: <code>internal-projectDir.R</code>, <code>internal-readSampleMetadataFile.R</code>, <code>internal-sampleDirs.R</code>. We may want to provide this code as a shared bcbio core package (e.g. bcbioBase) in the future.</li>
<li>Added internal utility to check for valid marker genes (<code>.validMarkers()</code>).</li>
<li>Improved Ensembl release version support (<code>ensemblVersion</code>).</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-18-2017-09-17" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-18-2017-09-17" class="anchor"></a>bcbioSingleCell 0.0.18 (2017-09-17)</h1>
<ul>
<li>Renamed <code><a href="../reference/deprecated.html">plotClusters()</a></code> to <code><a href="../reference/plotMarkers.html">plotMarkers()</a></code>. Added soft deprecation.</li>
<li>Added <a href="https://cran.r-project.org/web/packages/viridis/index.html">viridis</a> color support in tSNE plots and heatmaps.</li>
<li>Converted <code><a href="../reference/deprecated.html">loadSingleCellRun()</a></code> and <code><a href="../reference/loadCellRanger.html">loadCellRanger()</a></code> from S4 generics back to standard functions.</li>
<li>Added t-SNE utility functions: <code><a href="../reference/fetchTSNEData.html">fetchTSNEData()</a></code>, <code><a href="../reference/fetchTSNEExpressionData.html">fetchTSNEExpressionData()</a></code>, and <code><a href="../reference/deprecated.html">plotTSNEExpressionData()</a></code>. This enable plotting of geometric mean values of desired marker genes.</li>
<li>Updated NEWS to Markdown, with hyperlinks.</li>
<li>Offloaded generics that would otherwise conflict with bcbioRNASeq to the basejump package.</li>
<li>Improved roxygen documentation. Moved as much documentation as possible to the methods files.</li>
<li>Updated <code>cellCycleMarkers</code> and <code>cellTypeMarkers</code> data. Now supports Drosophila.</li>
<li>Sample IDs are now sanitized using <code>make.names()</code> instead of <code>camel()</code>. This avoids undesirable coercion of some IDs (e.g. <code>group1_1</code> into <code>group11</code>).</li>
<li>Added recommended package syntax guidelines.</li>
<li>lintr checks now allow implicit integers (e.g. <code>1</code> instead of <code>1L</code>).</li>
<li>Added Seurat as dependency in <code>DESCRIPTION</code> file. The package now attaches Seurat automatically.</li>
<li>Package no longer imports mononcle or suggests scater, scde, or scone. We’re planning on adding these back in a future update, but build checks on Travis CI otherwise take too long.</li>
<li>Added new <code><a href="../reference/quantileHeatmap.html">quantileHeatmap()</a></code> function.</li>
<li>Improved Markdown header support across functions, where applicable.</li>
<li>Improved <code>bcbioSCFiltered</code> to <code>seurat</code> coercion to slot relevant bcbio metadata.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-17-2017-09-03" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-17-2017-09-03" class="anchor"></a>bcbioSingleCell 0.0.17 (2017-09-03)</h1>
<ul>
<li>Renamed package from <code>bcbioSinglecell</code> to <code>bcbioSingleCell</code>.</li>
<li>Added <a href="https://cran.r-project.org/web/packages/viridis/index.html">viridis</a> color palette support to quality control plots.</li>
<li>Added cell-cycle marker genes for human and mouse.</li>
<li>Slotted <code>organism</code> in <code>bcbioSCDataSet</code> metadata, in addition to <code>genomeBuild</code>.</li>
<li>Updated <code>bcbioSCFiltered</code> to <code>seurat</code> coercion method to also run <code>FindVariableGenes()</code> and <code>ScaleData()</code> by default.</li>
<li>Added cell-cycle regression into <a href="http://satijalab.org/seurat">Seurat</a> clustering <a href="http://rmarkdown.rstudio.com">RMarkdown</a> template.</li>
<li>Improved <a href="https://github.com/hadley/pkgdown">pkgdown</a> settings and website appearance.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-16-2017-08-25" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-16-2017-08-25" class="anchor"></a>bcbioSingleCell 0.0.16 (2017-08-25)</h1>
<ul>
<li>Support for CRAN release of <a href="http://satijalab.org/seurat">Seurat</a>.</li>
<li>Improved documentation of package NAMESPACE in <code>bcbioSinglecell-package.R</code> file.</li>
<li>Offloaded <code>download()</code> functionality to <a href="http://steinbaugh.com/basejump">basejump</a> package, to avoid collisions with <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> package. Function has been renamed to <code>externalFile()</code>.</li>
<li>Removed function deprecations to simplify the NAMESPACE.</li>
<li>Improved <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> sample matching. With these changes the internal <code>.detectPipeline()</code> function is no longer needed.</li>
<li>Improved sample directory matching in internal <code>.sampleDirs()</code> function.</li>
<li>Updated paths to <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> MatrixMarket files in internal <code>.readSparseCounts()</code> function.</li>
<li>Updated use of <code>packageSE()</code> to <code>prepareSE()</code>, matching the corresponding <a href="http://steinbaugh.com/basejump">basejump</a> function change.</li>
<li>Renamed internal use of <code>filter()</code> to <code>tidy_filter()</code>, to avoid future NAMESPACE collisions with <a href="http://bioconductor.org/packages/release/bioc/html/ensembldb.html">ensembldb</a> package.</li>
<li>Renamed <code>bcbioSCSubset</code> class to <code>bcbioSCFiltered</code> class.</li>
<li>Fixed memory issue in <code>plotZeroesVsDepth()</code> for datasets with high cell counts.</li>
<li>Improved sample matching for <code><a href="../reference/selectSamples.html">selectSamples()</a></code>. Will attempt to migrate this to bracket-based subsetting in a future update.</li>
<li>Restricted sample selection with <code><a href="../reference/selectSamples.html">selectSamples()</a></code> to only work on <code>bcbioSCFiltered</code> class for the time being. We can add bracket-based subsetting or S4 method support in <code><a href="../reference/selectSamples.html">selectSamples()</a></code> to properly work on <code>bcbioSCDataSet</code> class in a future update.</li>
<li>Updated <a href="http://rmarkdown.rstudio.com">RMarkdown</a> template settings and simplified the setup chunks.</li>
<li>Initial support for <code>bcbioSCFiltered</code> class coercion to <a href="http://cole-trapnell-lab.github.io/monocle-release/">monocle</a> <code>CellDataSet</code> class.</li>
<li>Suggest <a href="http://bioconductor.org/packages/release/bioc/html/scater.html">scater</a> and <a href="https://bioconductor.org/packages/release/bioc/html/scone.html">scone</a> packages for additional quality control and visualization.</li>
<li>Require <a href="http://cole-trapnell-lab.github.io/monocle-release/">monocle</a> &gt;= 2.5.0.</li>
<li>Added initial support for <a href="https://cran.r-project.org/web/packages/viridis/index.html">viridis</a> color palette in quality control plots.</li>
<li>Suggest <a href="http://rmarkdown.rstudio.com">rmarkdown</a> and <a href="http://bioconductor.org/packages/release/bioc/html/scde.html">scde</a> packages.</li>
<li>Initial commit of <code><a href="../reference/subsetPerSample.html">subsetPerSample()</a></code> function.</li>
<li>Miscellaneous <a href="http://rmarkdown.rstudio.com">RMarkdown</a> template improvements.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-15-2017-08-11" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-15-2017-08-11" class="anchor"></a>bcbioSingleCell 0.0.15 (2017-08-11)</h1>
<ul>
<li>Renamed functions in <code>lowerCamelCase</code> from <code>snake_case</code>.</li>
<li>Draft support for <a href="http://cole-trapnell-lab.github.io/monocle-release/">monocle</a>, <a href="http://bioconductor.org/packages/release/bioc/html/scater.html">scater</a>, and <a href="https://bioconductor.org/packages/release/bioc/html/scone.html">scone</a>.</li>
<li>Package now depends on <a href="https://www.bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment</a>.</li>
<li>Renamed <code>loadRun()</code> to <code><a href="../reference/deprecated.html">loadSingleCellRun()</a></code> for improved compatibility with <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> package. This helps avoid NAMESPACE collisions between packages.</li>
<li>Improved support for custom GTF files.</li>
<li>Split out <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> import into a separate utility function named <code><a href="../reference/loadCellRanger.html">loadCellRanger()</a></code>.</li>
<li>Added plotZerosVsDepth() by @roryk.</li>
<li>Renamed <code>filteringCriteria</code> to <code>filterParams</code> in <code>@metadata</code> slot.</li>
<li>Miscellaneous documentation fixes.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-14-2017-07-26" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-14-2017-07-26" class="anchor"></a>bcbioSingleCell 0.0.14 (2017-07-26)</h1>
<ul>
<li>Migrated plotting functions to S4 methods.</li>
<li>Improved sample metadata handling.</li>
<li>Updated sample name and cellular barcode sanization.</li>
<li>Initial commit of <a href="http://satijalab.org/seurat">Seurat</a> utility functions.</li>
<li>Initial commit of YAML parameters for <a href="http://rmarkdown.rstudio.com">RMarkdown</a> templates.</li>
<li>Draft support for <a href="http://satijalab.org/seurat">Seurat</a> v2 pre-release.</li>
<li>Initial support for <a href="https://www.illumina.com/products/by-type/sequencing-kits/library-prep-kits/surecell-wta-ddseq.html">SureCell</a> UMIs.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-13-2017-07-10" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-13-2017-07-10" class="anchor"></a>bcbioSingleCell 0.0.13 (2017-07-10)</h1>
<ul>
<li>Integrated [bcbio][] and <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> workflows into <code>load_run()</code>.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-12-2017-06-28" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-12-2017-06-28" class="anchor"></a>bcbioSingleCell 0.0.12 (2017-06-28)</h1>
<ul>
<li>Initial support for S4 object creation using <a href="https://www.bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment</a>.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-11-2017-06-24" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-11-2017-06-24" class="anchor"></a>bcbioSingleCell 0.0.11 (2017-06-24)</h1>
<ul>
<li>Added <a href="https://github.com/hadley/testthat">testthat</a>, <a href="https://github.com/jimhester/lintr">lintr</a>, and <a href="https://github.com/jimhester/covr">covr</a> support for code coverage.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-10-2017-06-15" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-10-2017-06-15" class="anchor"></a>bcbioSingleCell 0.0.10 (2017-06-15)</h1>
<ul>
<li>Updated <a href="http://rmarkdown.rstudio.com">RMarkdown</a> templates.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-9-2017-06-12" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-9-2017-06-12" class="anchor"></a>bcbioSingleCell 0.0.9 (2017-06-12)</h1>
<ul>
<li>Compatibility update for basejump S4 NAMESPACE changes.</li>
<li>Improved plots for quality control and filtering.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-8-2017-05-28" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-8-2017-05-28" class="anchor"></a>bcbioSingleCell 0.0.8 (2017-05-28)</h1>
<ul>
<li>Renamed mitochondrial plot functions.</li>
<li>Changed presentation of mitochondrial abundance as ratio instead of percentage.</li>
<li>Simplified NAMESPACE by offloading dependencies to basejump.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-7-2017-05-13" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-7-2017-05-13" class="anchor"></a>bcbioSingleCell 0.0.7 (2017-05-13)</h1>
<ul>
<li>Initial support for loading of 10x Genomics <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> output.</li>
<li>Add detection of droplet method based on the metadata file.</li>
<li>Draft support for import of <a href="https://1cell-bio.com">inDrop</a> i5 index barcode counts.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-6-2017-05-10" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-6-2017-05-10" class="anchor"></a>bcbioSingleCell 0.0.6 (2017-05-10)</h1>
<ul>
<li>Modified <code>load_run()</code> function for improved consistency with <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> package.</li>
<li>Initial incorporation of scater package into workflow.</li>
<li>Initial commit of RMarkdown template skeletons.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-5-2017-05-08" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-5-2017-05-08" class="anchor"></a>bcbioSingleCell 0.0.5 (2017-05-08)</h1>
<ul>
<li>Improvements to run object based on code in <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> package.</li>
<li>Better handling in plots for datasets with many samples.</li>
<li>Labeled quality control cutoffs on the plots.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-4-2017-04-26" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-4-2017-04-26" class="anchor"></a>bcbioSingleCell 0.0.4 (2017-04-26)</h1>
<ul>
<li>Prepared package for <a href="http://www.tidyverse.org">tidyverse</a> dependency updates.</li>
<li>Updated HPC server detection.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-3-2017-04-20" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-3-2017-04-20" class="anchor"></a>bcbioSingleCell 0.0.3 (2017-04-20)</h1>
<ul>
<li>NAMESPACE consolidation and function cleanup.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-2-2017-04-12" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-2-2017-04-12" class="anchor"></a>bcbioSingleCell 0.0.2 (2017-04-12)</h1>
<ul>
<li>Added quality control plotting functions.</li>
<li>Converted all functions to standard evaluation.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-1-2017-03-01" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-1-2017-03-01" class="anchor"></a>bcbioSingleCell 0.0.1 (2017-03-01)</h1>
<ul>
<li>Initial draft release.</li>
</ul>
</div>
    </div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#bcbiosinglecell-0-0-22-2017-11-17">0.0.22</a></li>
        <li><a href="#bcbiosinglecell-0-0-21-2017-11-08">0.0.21</a></li>
        <li><a href="#bcbiosinglecell-0-0-20-2017-10-24">0.0.20</a></li>
        <li><a href="#bcbiosinglecell-0-0-19-2017-10-12">0.0.19</a></li>
        <li><a href="#bcbiosinglecell-0-0-18-2017-09-17">0.0.18</a></li>
        <li><a href="#bcbiosinglecell-0-0-17-2017-09-03">0.0.17</a></li>
        <li><a href="#bcbiosinglecell-0-0-16-2017-08-25">0.0.16</a></li>
        <li><a href="#bcbiosinglecell-0-0-15-2017-08-11">0.0.15</a></li>
        <li><a href="#bcbiosinglecell-0-0-14-2017-07-26">0.0.14</a></li>
        <li><a href="#bcbiosinglecell-0-0-13-2017-07-10">0.0.13</a></li>
        <li><a href="#bcbiosinglecell-0-0-12-2017-06-28">0.0.12</a></li>
        <li><a href="#bcbiosinglecell-0-0-11-2017-06-24">0.0.11</a></li>
        <li><a href="#bcbiosinglecell-0-0-10-2017-06-15">0.0.10</a></li>
        <li><a href="#bcbiosinglecell-0-0-9-2017-06-12">0.0.9</a></li>
        <li><a href="#bcbiosinglecell-0-0-8-2017-05-28">0.0.8</a></li>
        <li><a href="#bcbiosinglecell-0-0-7-2017-05-13">0.0.7</a></li>
        <li><a href="#bcbiosinglecell-0-0-6-2017-05-10">0.0.6</a></li>
        <li><a href="#bcbiosinglecell-0-0-5-2017-05-08">0.0.5</a></li>
        <li><a href="#bcbiosinglecell-0-0-4-2017-04-26">0.0.4</a></li>
        <li><a href="#bcbiosinglecell-0-0-3-2017-04-20">0.0.3</a></li>
        <li><a href="#bcbiosinglecell-0-0-2-2017-04-12">0.0.2</a></li>
        <li><a href="#bcbiosinglecell-0-0-1-2017-03-01">0.0.1</a></li>
      </ul>
    </div>
  </div>

</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='http://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
