<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>All news • bcbioSingleCell</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->
<link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/readable/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">


<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>
  <link href="../extra.css" rel="stylesheet">
  
<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-news">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">bcbioSingleCell</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="..//index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioSingleCell">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

      <div class="row">

  <div class="col-md-9">
    <div class="page-header">
      <h1>Change log <small>All releases</small></h1>
    </div>

    <div class="contents">
    <div id="bcbiosinglecell-0-0-19" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-19" class="anchor"></a>bcbioSingleCell 0.0.19</h1>
<ul>
<li>Renamed main object class from <code>bcbioSCDataSet</code> to <code>bcbioSingleCell</code>.</li>
<li>Cell filtering with <code><a href="../reference/filterCells.html">filterCells()</a></code> will now slot a named logical vector into <code>metadata(object)[["filteredCells"]]</code>, which will be used to dynamically subset the slotted internal <code>SummarizedExperiment</code> data. Now that we’re using this approach, we can return a modified <code>bcbioSingleCell</code> object rather than defining a separate <code>bcbioSCFiltered</code> class.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-18" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-18" class="anchor"></a>bcbioSingleCell 0.0.18</h1>
<ul>
<li>Renamed <code><a href="../reference/deprecated.html">plotClusters()</a></code> to <code><a href="../reference/plotMarkers.html">plotMarkers()</a></code>. Added soft deprecation.</li>
<li>Added <a href="https://cran.r-project.org/web/packages/viridis/index.html">viridis</a> color support in tSNE plots and heatmaps.</li>
<li>Converted <code><a href="../reference/deprecated.html">loadSingleCellRun()</a></code> and <code><a href="../reference/loadCellRanger.html">loadCellRanger()</a></code> from S4 generics back to standard functions.</li>
<li>Added t-SNE utility functions: <code><a href="../reference/fetchTSNEData.html">fetchTSNEData()</a></code>, <code><a href="../reference/fetchTSNEExpressionData.html">fetchTSNEExpressionData()</a></code>, and <code><a href="../reference/deprecated.html">plotTSNEExpressionData()</a></code>. This enable plotting of geometric mean values of desired marker genes.</li>
<li>Updated NEWS to Markdown, with hyperlinks.</li>
<li>Offloaded generics that would otherwise conflict with bcbioRNASeq to the basejump package.</li>
<li>Improved roxygen documentation. Moved as much documentation as possible to the methods files.</li>
<li>Updated <code>cellCycleMarkers</code> and <code>cellTypeMarkers</code> data. Now supports Drosophila.</li>
<li>Sample IDs are now sanitized using <code>make.names()</code> instead of <code>camel()</code>. This avoids undesirable coercion of some IDs (e.g. <code>group1_1</code> into <code>group11</code>).</li>
<li>Added recommended package syntax guidelines.</li>
<li>lintr checks now allow implicit integers (e.g. <code>1</code> instead of <code>1L</code>).</li>
<li>Added Seurat as dependency in <code>DESCRIPTION</code> file. The package now attaches Seurat automatically.</li>
<li>Package no longer imports mononcle or suggests scater, scde, or scone. We’re planning on adding these back in a future update, but build checks on Travis CI otherwise take too long.</li>
<li>Added new <code><a href="../reference/quantileHeatmap.html">quantileHeatmap()</a></code> function.</li>
<li>Improved Markdown header support across functions, where applicable.</li>
<li>Improved <code>bcbioSCFiltered</code> to <code>seurat</code> coercion to slot relevant bcbio metadata.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-17" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-17" class="anchor"></a>bcbioSingleCell 0.0.17</h1>
<ul>
<li>Renamed package from <code>bcbioSinglecell</code> to <code>bcbioSingleCell</code>.</li>
<li>Added <a href="https://cran.r-project.org/web/packages/viridis/index.html">viridis</a> color palette support to quality control plots.</li>
<li>Added cell-cycle marker genes for human and mouse.</li>
<li>Slotted <code>organism</code> in <code>bcbioSCDataSet</code> metadata, in addition to <code>genomeBuild</code>.</li>
<li>Updated <code>bcbioSCFiltered</code> to <code>seurat</code> coercion method to also run <code>FindVariableGenes()</code> and <code>ScaleData()</code> by default.</li>
<li>Added cell-cycle regression into <a href="http://satijalab.org/seurat">Seurat</a> clustering <a href="http://rmarkdown.rstudio.com">RMarkdown</a> template.</li>
<li>Improved <a href="https://github.com/hadley/pkgdown">pkgdown</a> settings and website appearance.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-16" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-16" class="anchor"></a>bcbioSingleCell 0.0.16</h1>
<ul>
<li>Support for CRAN release of <a href="http://satijalab.org/seurat">Seurat</a>.</li>
<li>Improved documentation of package NAMESPACE in <code>bcbioSinglecell-package.R</code> file.</li>
<li>Offloaded <code>download()</code> functionality to <a href="http://steinbaugh.com/basejump">basejump</a> package, to avoid collisions with <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> package. Function has been renamed to <code>externalFile()</code>.</li>
<li>Removed function deprecations to simplify the NAMESPACE.</li>
<li>Improved <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> sample matching. With these changes the internal <code>.detectPipeline()</code> function is no longer needed.</li>
<li>Improved sample directory matching in internal <code>.sampleDirs()</code> function.</li>
<li>Updated paths to <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> MatrixMarket files in internal <code>.readSparseCounts()</code> function.</li>
<li>Updated use of <code>packageSE()</code> to <code>prepareSE()</code>, matching the corresponding <a href="http://steinbaugh.com/basejump">basejump</a> function change.</li>
<li>Renamed internal use of <code>filter()</code> to <code>tidy_filter()</code>, to avoid future NAMESPACE collisions with <a href="http://bioconductor.org/packages/release/bioc/html/ensembldb.html">ensembldb</a> package.</li>
<li>Renamed <code>bcbioSCSubset</code> class to <code>bcbioSCFiltered</code> class.</li>
<li>Fixed memory issue in <code>plotZeroesVsDepth()</code> for datasets with high cell counts.</li>
<li>Improved sample matching for <code><a href="../reference/selectSamples.html">selectSamples()</a></code>. Will attempt to migrate this to bracket-based subsetting in a future update.</li>
<li>Restricted sample selection with <code><a href="../reference/selectSamples.html">selectSamples()</a></code> to only work on <code>bcbioSCFiltered</code> class for the time being. We can add bracket-based subsetting or S4 method support in <code><a href="../reference/selectSamples.html">selectSamples()</a></code> to properly work on <code>bcbioSCDataSet</code> class in a future update.</li>
<li>Updated <a href="http://rmarkdown.rstudio.com">RMarkdown</a> template settings and simplified the setup chunks.</li>
<li>Initial support for <code>bcbioSCFiltered</code> class coercion to <a href="http://cole-trapnell-lab.github.io/monocle-release/">monocle</a> <code>CellDataSet</code> class.</li>
<li>Suggest <a href="http://bioconductor.org/packages/release/bioc/html/scater.html">scater</a> and <a href="https://bioconductor.org/packages/release/bioc/html/scone.html">scone</a> packages for additional quality control and visualization.</li>
<li>Require <a href="http://cole-trapnell-lab.github.io/monocle-release/">monocle</a> &gt;= 2.5.0.</li>
<li>Added initial support for <a href="https://cran.r-project.org/web/packages/viridis/index.html">viridis</a> color palette in quality control plots.</li>
<li>Suggest <a href="http://rmarkdown.rstudio.com">rmarkdown</a> and <a href="http://bioconductor.org/packages/release/bioc/html/scde.html">scde</a> packages.</li>
<li>Initial commit of <code><a href="../reference/subsetPerSample.html">subsetPerSample()</a></code> function.</li>
<li>Miscellaneous <a href="http://rmarkdown.rstudio.com">RMarkdown</a> template improvements.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-15" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-15" class="anchor"></a>bcbioSingleCell 0.0.15</h1>
<ul>
<li>Renamed functions in <code>lowerCamelCase</code> from <code>snake_case</code>.</li>
<li>Draft support for <a href="http://cole-trapnell-lab.github.io/monocle-release/">monocle</a>, <a href="http://bioconductor.org/packages/release/bioc/html/scater.html">scater</a>, and <a href="https://bioconductor.org/packages/release/bioc/html/scone.html">scone</a>.</li>
<li>Package now depends on <a href="https://www.bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment</a>.</li>
<li>Renamed <code>loadRun()</code> to <code><a href="../reference/deprecated.html">loadSingleCellRun()</a></code> for improved compatibility with <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> package. This helps avoid NAMESPACE collisions between packages.</li>
<li>Improved support for custom GTF files.</li>
<li>Split out <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> import into a separate utility function named <code><a href="../reference/loadCellRanger.html">loadCellRanger()</a></code>.</li>
<li>Added plotZerosVsDepth() by @roryk.</li>
<li>Renamed <code>filteringCriteria</code> to <code>filterParams</code> in <code>@metadata</code> slot.</li>
<li>Miscellaneous documentation fixes.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-14" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-14" class="anchor"></a>bcbioSingleCell 0.0.14</h1>
<ul>
<li>Migrated plotting functions to S4 methods.</li>
<li>Improved sample metadata handling.</li>
<li>Updated sample name and cellular barcode sanization.</li>
<li>Initial commit of <a href="http://satijalab.org/seurat">Seurat</a> utility functions.</li>
<li>Initial commit of YAML parameters for <a href="http://rmarkdown.rstudio.com">RMarkdown</a> templates.</li>
<li>Draft support for <a href="http://satijalab.org/seurat">Seurat</a> v2 pre-release.</li>
<li>Initial support for <a href="https://www.illumina.com/products/by-type/sequencing-kits/library-prep-kits/surecell-wta-ddseq.html">SureCell</a> UMIs.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-13" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-13" class="anchor"></a>bcbioSingleCell 0.0.13</h1>
<ul>
<li>Integrated [bcbio][] and <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> workflows into <code>load_run()</code>.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-12" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-12" class="anchor"></a>bcbioSingleCell 0.0.12</h1>
<ul>
<li>Initial support for S4 object creation using <a href="https://www.bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment</a>.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-11" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-11" class="anchor"></a>bcbioSingleCell 0.0.11</h1>
<ul>
<li>Added <a href="https://github.com/hadley/testthat">testthat</a>, <a href="https://github.com/jimhester/lintr">lintr</a>, and <a href="https://github.com/jimhester/covr">covr</a> support for code coverage.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-10" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-10" class="anchor"></a>bcbioSingleCell 0.0.10</h1>
<ul>
<li>Updated <a href="http://rmarkdown.rstudio.com">RMarkdown</a> templates.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-9" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-9" class="anchor"></a>bcbioSingleCell 0.0.9</h1>
<ul>
<li>Compatibility update for basejump S4 NAMESPACE changes.</li>
<li>Improved plots for quality control and filtering.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-8" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-8" class="anchor"></a>bcbioSingleCell 0.0.8</h1>
<ul>
<li>Renamed mitochondrial plot functions.</li>
<li>Changed presentation of mitochondrial abundance as ratio instead of percentage.</li>
<li>Simplified NAMESPACE by offloading dependencies to basejump.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-7" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-7" class="anchor"></a>bcbioSingleCell 0.0.7</h1>
<ul>
<li>Initial support for loading of 10x Genomics <a href="https://support.10xgenomics.com/single-cell-gene-expression/software/pipelines/latest/what-is-cell-ranger">Cell Ranger</a> output.</li>
<li>Add detection of droplet method based on the metadata file.</li>
<li>Draft support for import of <a href="https://1cell-bio.com">inDrop</a> i5 index barcode counts.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-6" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-6" class="anchor"></a>bcbioSingleCell 0.0.6</h1>
<ul>
<li>Modified <code>load_run()</code> function for improved consistency with <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> package.</li>
<li>Initial incorporation of scater package into workflow.</li>
<li>Initial commit of RMarkdown template skeletons.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-5" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-5" class="anchor"></a>bcbioSingleCell 0.0.5</h1>
<ul>
<li>Improvements to run object based on code in <a href="http://bioinformatics.sph.harvard.edu/bcbioRNASeq">bcbioRNASeq</a> package.</li>
<li>Better handling in plots for datasets with many samples.</li>
<li>Labeled quality control cutoffs on the plots.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-4" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-4" class="anchor"></a>bcbioSingleCell 0.0.4</h1>
<ul>
<li>Prepared package for <a href="http://www.tidyverse.org">tidyverse</a> dependency updates.</li>
<li>Updated HPC server detection.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-3" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-3" class="anchor"></a>bcbioSingleCell 0.0.3</h1>
<ul>
<li>NAMESPACE consolidation and function cleanup.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-2" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-2" class="anchor"></a>bcbioSingleCell 0.0.2</h1>
<ul>
<li>Added quality control plotting functions.</li>
<li>Converted all functions to standard evaluation.</li>
</ul>
</div>
    <div id="bcbiosinglecell-0-0-1" class="section level1">
<h1 class="hasAnchor">
<a href="#bcbiosinglecell-0-0-1" class="anchor"></a>bcbioSingleCell 0.0.1</h1>
<ul>
<li>Initial draft release.</li>
</ul>
</div>
    </div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div id="tocnav">
      <h2>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
        <li><a href="#bcbiosinglecell-0-0-19">0.0.19</a></li>
        <li><a href="#bcbiosinglecell-0-0-18">0.0.18</a></li>
        <li><a href="#bcbiosinglecell-0-0-17">0.0.17</a></li>
        <li><a href="#bcbiosinglecell-0-0-16">0.0.16</a></li>
        <li><a href="#bcbiosinglecell-0-0-15">0.0.15</a></li>
        <li><a href="#bcbiosinglecell-0-0-14">0.0.14</a></li>
        <li><a href="#bcbiosinglecell-0-0-13">0.0.13</a></li>
        <li><a href="#bcbiosinglecell-0-0-12">0.0.12</a></li>
        <li><a href="#bcbiosinglecell-0-0-11">0.0.11</a></li>
        <li><a href="#bcbiosinglecell-0-0-10">0.0.10</a></li>
        <li><a href="#bcbiosinglecell-0-0-9">0.0.9</a></li>
        <li><a href="#bcbiosinglecell-0-0-8">0.0.8</a></li>
        <li><a href="#bcbiosinglecell-0-0-7">0.0.7</a></li>
        <li><a href="#bcbiosinglecell-0-0-6">0.0.6</a></li>
        <li><a href="#bcbiosinglecell-0-0-5">0.0.5</a></li>
        <li><a href="#bcbiosinglecell-0-0-4">0.0.4</a></li>
        <li><a href="#bcbiosinglecell-0-0-3">0.0.3</a></li>
        <li><a href="#bcbiosinglecell-0-0-2">0.0.2</a></li>
        <li><a href="#bcbiosinglecell-0-0-1">0.0.1</a></li>
      </ul>
    </div>
  </div>

</div>

      <footer>
      <div class="copyright">
  <p>Developed by Michael Steinbaugh, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  </body>
</html>
