<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>bcbio Single-Cell RNA-Seq • bcbioSingleCell</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><link href="extra.css" rel="stylesheet">
<meta property="og:title" content="bcbio Single-Cell RNA-Seq">
<meta property="og:description" content="R package for bcbio single-cell RNA-seq analysis.">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">bcbioSingleCell</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.4.10</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="index.html">
    <span class="fas fa fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li>
  <a href="news/index.html">Changelog</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/hbc/bcbioSingleCell">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div id="bcbiosinglecell" class="section level1">
<div class="page-header"><h1 class="hasAnchor">
<a href="#bcbiosinglecell" class="anchor"></a>bcbioSingleCell</h1></div>

<p><a href="https://www.r-project.org/">R</a> package for <a href="https://bcbio-nextgen.readthedocs.io/">bcbio</a> single-cell RNA-seq analysis.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div id="r-method" class="section level3">
<h3 class="hasAnchor">
<a href="#r-method" class="anchor"></a><a href="https://www.r-project.org/">R</a> method</h3>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"remotes"</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)) {
    <span class="kw"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"remotes"</span>)
}
<span class="kw"><a href="https://rdrr.io/r/base/Sys.setenv.html">Sys.setenv</a></span>(<span class="dt">R_REMOTES_UPGRADE =</span> <span class="st">"always"</span>)
<span class="co"># Set `GITHUB_PAT` in `~/.Renviron` if you get a rate limit error.</span>
remotes<span class="op">::</span><span class="kw"><a href="https://remotes.r-lib.org/reference/install_github.html">install_github</a></span>(<span class="st">"hbc/bcbioSingleCell"</span>)</code></pre></div>
<p>Here’s how to update to the latest version on GitHub:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/Sys.setenv.html">Sys.setenv</a></span>(<span class="dt">R_REMOTES_UPGRADE =</span> <span class="st">"always"</span>)
remotes<span class="op">::</span><span class="kw"><a href="https://remotes.r-lib.org/reference/update_packages.html">update_packages</a></span>()</code></pre></div>
<p>Always check that your Bioconductor installation is valid before proceeding.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="cf">if</span> (<span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span>(<span class="st">"BiocManager"</span>, <span class="dt">quietly =</span> <span class="ot">TRUE</span>)) {
    <span class="kw"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span>(<span class="st">"BiocManager"</span>)
}
BiocManager<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/BiocManager/man/valid.html">valid</a></span>()</code></pre></div>
</div>
<div id="conda-method" class="section level3">
<h3 class="hasAnchor">
<a href="#conda-method" class="anchor"></a><a href="https://conda.io/">Conda</a> method</h3>
<p>Configure <a href="https://conda.io/">Conda</a> to use the <a href="https://bioconda.github.io/">Bioconda</a> channels.</p>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="co"># Don't install recipe into base environment.</span>
<span class="va">name=</span><span class="st">"r-bcbiosinglecell"</span>
<span class="ex">conda</span> create --name=<span class="st">"</span><span class="va">$name</span><span class="st">"</span> <span class="st">"</span><span class="va">$name</span><span class="st">"</span>
<span class="ex">conda</span> activate <span class="st">"</span><span class="va">$name</span><span class="st">"</span>
<span class="ex">R</span></code></pre></div>
</div>
<div id="docker-method" class="section level3">
<h3 class="hasAnchor">
<a href="#docker-method" class="anchor"></a><a href="https://www.docker.com/">Docker</a> method</h3>
<div class="sourceCode"><pre class="sourceCode sh"><code class="sourceCode bash"><span class="va">image=</span><span class="st">"acidgenomics/bcbiosinglecell"</span>
<span class="va">workdir=</span><span class="st">"/work"</span>
<span class="ex">docker</span> pull <span class="st">"</span><span class="va">$image</span><span class="st">"</span>
<span class="ex">docker</span> run -it \
    --volume=<span class="st">"</span><span class="va">${PWD}</span><span class="st">:</span><span class="va">${workdir}</span><span class="st">"</span> \
    --workdir=<span class="st">"</span><span class="va">$workdir</span><span class="st">"</span> \
    <span class="st">"</span><span class="va">$image</span><span class="st">"</span> \
    R</code></pre></div>
</div>
</div>
<div id="load-bcbio-single-cell-rna-seq-data" class="section level2">
<h2 class="hasAnchor">
<a href="#load-bcbio-single-cell-rna-seq-data" class="anchor"></a>Load <a href="https://bcbio-nextgen.readthedocs.io/">bcbio</a> single-cell RNA-seq data</h2>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(bcbioSingleCell)
bcb &lt;-<span class="st"> </span><span class="kw"><a href="reference/bcbioSingleCell.html">bcbioSingleCell</a></span>(
    <span class="dt">uploadDir =</span> <span class="st">"indrops/final"</span>,
    <span class="dt">interestingGroups =</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"genotype"</span>, <span class="st">"treatment"</span>),
    <span class="dt">sampleMetadataFile =</span> <span class="st">"sample_metadata.csv"</span>,
    <span class="dt">organism =</span> <span class="st">"Homo sapiens"</span>,
    <span class="dt">ensemblRelease =</span> 90L
)
<span class="kw">saveData</span>(bcb, <span class="dt">dir =</span> <span class="st">"."</span>)</code></pre></div>
<p>This will return a <code>bcbioSingleCell</code> object, which is an extension of the <a href="https://bioconductor.org/">Bioconductor</a> <a href="https://bioconductor.org/packages/SingleCellExperiment/">SingleCellExperiment</a> container class. Consult the <code><a href="reference/bcbioSingleCell.html">bcbioSingleCell()</a></code> constructor function documentation for detailed information on the supported parameters:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/utils/help.html">help</a></span>(<span class="dt">topic =</span> <span class="st">"bcbioSingleCell"</span>, <span class="dt">package =</span> <span class="st">"bcbioSingleCell"</span>)</code></pre></div>
</div>
<div id="sample-metadata-examples" class="section level2">
<h2 class="hasAnchor">
<a href="#sample-metadata-examples" class="anchor"></a>Sample metadata examples</h2>
<div id="fastq-files-with-samples-multiplexed-by-index-barcode" class="section level3">
<h3 class="hasAnchor">
<a href="#fastq-files-with-samples-multiplexed-by-index-barcode" class="anchor"></a>FASTQ files with samples multiplexed by index barcode</h3>
<p>This is our current recommended method for analyzing an <a href="https://github.com/indrops/indrops/">inDrops</a> dataset. The sample index barcodes are multiplexed per FASTQ set. For Illumina sequencing data, the raw binary base call (BCL) data must be converted into FASTQs (split into <code>R1</code>-<code>R4</code> files) using <a href="https://support.illumina.com/sequencing/sequencing_software/bcl2fastq-conversion-software.html">bcl2fastq</a>.</p>
<p>The <a href="https://github.com/indrops/indrops/">inDrops</a> library version is automatically detected by <a href="https://bcbio-nextgen.readthedocs.io/">bcbio</a>, but ensure that the sample index sequences provided match the library version when attempting to create a <code>bcbioSingleCell</code> object.</p>
<p>Consult the <a href="https://bcbio-nextgen.readthedocs.io/">bcbio</a> documentation for more information on how to configure an <a href="https://github.com/indrops/indrops/">inDrops</a> run prior to loading into <a href="https://www.r-project.org/">R</a> with the <code><a href="reference/bcbioSingleCell.html">bcbioSingleCell()</a></code> function.</p>
<table class="table">
<thead><tr class="header">
<th>description</th>
<th>index</th>
<th>sequence</th>
<th>sampleName</th>
<th>aggregate</th>
<th>genotype</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>indrops1</td>
<td>1</td>
<td>CTCTCTAT</td>
<td>sample1_1</td>
<td>sample1</td>
<td>wildtype</td>
</tr>
<tr class="even">
<td>indrops1</td>
<td>2</td>
<td>TATCCTCT</td>
<td>sample2_1</td>
<td>sample2</td>
<td>knockout</td>
</tr>
<tr class="odd">
<td>indrops1</td>
<td>3</td>
<td>GTAAGGAG</td>
<td>sample3_1</td>
<td>sample3</td>
<td>wildtype</td>
</tr>
<tr class="even">
<td>indrops1</td>
<td>4</td>
<td>ACTGCATA</td>
<td>sample4_1</td>
<td>sample4</td>
<td>knockout</td>
</tr>
<tr class="odd">
<td>indrops2</td>
<td>1</td>
<td>CTCTCTAT</td>
<td>sample1_2</td>
<td>sample1</td>
<td>wildtype</td>
</tr>
<tr class="even">
<td>indrops2</td>
<td>2</td>
<td>TATCCTCT</td>
<td>sample1_2</td>
<td>sample2</td>
<td>knockout</td>
</tr>
<tr class="odd">
<td>indrops2</td>
<td>3</td>
<td>GTAAGGAG</td>
<td>sample1_2</td>
<td>sample3</td>
<td>wildtype</td>
</tr>
<tr class="even">
<td>indrops2</td>
<td>4</td>
<td>ACTGCATA</td>
<td>sample1_2</td>
<td>sample4</td>
<td>knockout</td>
</tr>
</tbody>
</table>
<p>Note that bcbio currently outputs the reverse complement index sequence in the sample directory names (e.g. <code>sample-ATAGAGAG</code>). Define the forward index barcode in the <code>sequence</code> column here, not the reverse complement. The reverse complement will be calculated automatically and added as the <code>revcomp</code> column in the sample metadata.</p>
</div>
<div id="fastq-files-demultiplexed-per-sample" class="section level3">
<h3 class="hasAnchor">
<a href="#fastq-files-demultiplexed-per-sample" class="anchor"></a>FASTQ files demultiplexed per sample</h3>
<p>This is our current method for handling <a href="https://www.10xgenomics.com/solutions/single-cell/">10X Genomics Chromium</a> and <a href="https://www.illumina.com/products/by-type/sequencing-kits/library-prep-kits/surecell-wta-ddseq.html">Illumina SureCell</a> cell barcodes.</p>
<table class="table">
<thead><tr class="header">
<th>description</th>
<th>genotype</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>sample1</td>
<td>wildtype</td>
</tr>
<tr class="even">
<td>sample2</td>
<td>knockout</td>
</tr>
<tr class="odd">
<td>sample3</td>
<td>wildtype</td>
</tr>
<tr class="even">
<td>sample4</td>
<td>knockout</td>
</tr>
</tbody>
</table>
</div>
<div id="invalid-object" class="section level3">
<h3 class="hasAnchor">
<a href="#invalid-object" class="anchor"></a>Invalid object</h3>
<p>If you encounter a <code>validObject</code> error when attempting to load a <code>bcbioSingleCell</code> object from a previous analysis, run this step to update the object to the current version of the package:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">object &lt;-<span class="st"> </span><span class="kw"><a href="reference/updateObject.html">updateObject</a></span>(object)
<span class="kw">validObject</span>(object)
## [1] TRUE</code></pre></div>
</div>
</div>
<div id="references" class="section level2">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<p>The papers and software cited in our workflows are available as a <a href="https://paperpile.com/shared/C8EMxl">shared library</a> on <a href="https://paperpile.com/">Paperpile</a>.</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <div class="links">
<h2>Links</h2>
<ul class="list-unstyled">
<li>Browse source code at <br><a href="https://github.com/hbc/bcbioSingleCell">https://​github.com/​hbc/​bcbioSingleCell</a>
</li>
<li>Report a bug at <br><a href="https://github.com/hbc/bcbioSingleCell/issues/">https://​github.com/​hbc/​bcbioSingleCell/​issues/​</a>
</li>
</ul>
</div>
<div class="license">
<h2>License</h2>
<ul class="list-unstyled">
<li><a href="https://www.r-project.org/Licenses/GPL-3">GPL-3</a></li>
</ul>
</div>
<div class="community">
<h2>Community</h2>
<ul class="list-unstyled">
<li><a href="CONTRIBUTING.html">Contributing guide</a></li>
</ul>
</div>
<div class="developers">
<h2>Developers</h2>
<ul class="list-unstyled">
<li>
<a href="https://steinbaugh.com/">Michael Steinbaugh</a> <br><small class="roles"> Author, maintainer </small> <a href="https://orcid.org/0000-0002-2403-2221" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID"></a> </li>
<li>
<a href="https://github.com/roryk/">Rory Kirchner</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-4814-5885" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID"></a> </li>
<li>
<a href="https://github.com/marypiper/">Mary Piper</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-2699-3840" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID"></a> </li>
<li>
<a href="https://github.com/vbarrera/">Victor Barrera</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0003-0590-4634" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID"></a> </li>
<li>
<a href="https://github.com/sjhosui/">Shannan Ho Sui</a> <br><small class="roles"> Author </small> <a href="https://orcid.org/0000-0002-6191-4709" target="orcid.widget"><img src="https://members.orcid.org/sites/default/files/vector_iD_icon.svg" class="orcid" alt="ORCID"></a> </li>
<li>
<a href="https://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core</a> <br><small class="roles"> Copyright holder, funder </small>  </li>
</ul>
</div>

  <div class="dev-status">
<h2>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://www.repostatus.org/#active"><img src="https://www.repostatus.org/badges/latest/active.svg" alt="Repo status: active"></a></li>
<li><a href="https://travis-ci.org/hbc/bcbioSingleCell"><img src="https://travis-ci.org/hbc/bcbioSingleCell.svg?branch=master" alt="Travis CI build status"></a></li>
<li><a href="https://ci.appveyor.com/project/mjsteinbaugh/bcbiosinglecell/branch/master"><img src="https://ci.appveyor.com/api/projects/status/npy0mhfjn9saqv4g/branch/master?svg=true" alt="AppVeyor CI build status"></a></li>
<li><a href="https://anaconda.org/bioconda/r-bcbiosinglecell"><img src="https://anaconda.org/bioconda/r-bcbiosinglecell/badges/version.svg" alt="Anaconda version"><img src="https://anaconda.org/bioconda/r-bcbiosinglecell/badges/latest_release_date.svg" alt="Anaconda latest release date"><img src="https://anaconda.org/bioconda/r-bcbiosinglecell/badges/downloads.svg" alt="Anaconda downloads"></a></li>
</ul>
</div>
</div>
</div>


      <footer><div class="copyright">
  <p>Developed by <a href="https://steinbaugh.com/">Michael Steinbaugh</a>, <a href="https://github.com/roryk/">Rory Kirchner</a>, <a href="https://github.com/marypiper/">Mary Piper</a>, <a href="https://github.com/vbarrera/">Victor Barrera</a>, <a href="https://github.com/sjhosui/">Shannan Ho Sui</a>, <a href="https://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
