<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html lang="en">
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>bcbio Single-Cell RNA-Seq Data Set — bcbioSingleCell • bcbioSingleCell</title>

<!-- jquery -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha256-916EbMg70RQy9LHiGkXzG8hSg9EdNy97GazNG/aiY1w=" crossorigin="anonymous" />
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous" />

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script>

<!-- sticky kit -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="bcbio Single-Cell RNA-Seq Data Set — bcbioSingleCell" />

<meta property="og:description" content="bcbioSingleCell is an S4 class that extends SingleCellExperiment, and is
designed to store a bcbio single-cell RNA-seq analysis. This class contains
read counts saved as a sparse matrix (sparseMatrix), sample metadata, and
cell quality control metrics." />
<meta name="twitter:card" content="summary" />



<!-- mathjax -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioSingleCell</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.4</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioSingleCell">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>bcbio Single-Cell RNA-Seq Data Set</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioSingleCell/blob/master/R/bcbioSingleCell.R'><code>R/bcbioSingleCell.R</code></a></small>
    <div class="hidden name"><code>bcbioSingleCell.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p><code>bcbioSingleCell</code> is an S4 class that extends <code>SingleCellExperiment</code>, and is
designed to store a bcbio single-cell RNA-seq analysis. This class contains
read counts saved as a sparse matrix (<code>sparseMatrix</code>), sample metadata, and
cell quality control metrics.</p>
    
    </div>

    <pre class="usage"><span class='fu'>bcbioSingleCell</span>(<span class='no'>uploadDir</span>, <span class='kw'>sampleMetadataFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>organism</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>ensemblRelease</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>genomeBuild</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>gffFile</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>transgeneNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>spikeNames</span> <span class='kw'>=</span> <span class='kw'>NULL</span>,
  <span class='kw'>interestingGroups</span> <span class='kw'>=</span> <span class='st'>"sampleName"</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>uploadDir</th>
      <td><p><code>string</code>. Path to final upload directory. This path is set
when running "<code>bcbio_nextgen -w template</code>".</p></td>
    </tr>
    <tr>
      <th>sampleMetadataFile</th>
      <td><p><code>string</code>. Sample metadata file path. CSV
or TSV is preferred, but Excel worksheets are also supported. Check the
documentation for conventions and required columns.</p></td>
    </tr>
    <tr>
      <th>organism</th>
      <td><p><code>string</code>. Full Latin organism name (e.g. "<code>Homo sapiens</code>").</p></td>
    </tr>
    <tr>
      <th>ensemblRelease</th>
      <td><p><code>scalar integer</code>. Ensembl release version (e.g. <code>90</code>).
We recommend setting this value if possible, for improved reproducibility.
When left unset, the latest release available via AnnotationHub/ensembldb
is used. Note that the latest version available can vary, depending on the
versions of AnnotationHub and ensembldb in use.</p></td>
    </tr>
    <tr>
      <th>genomeBuild</th>
      <td><p><code>string</code>. Ensembl genome build assembly nam (e.g.
<code>"GRCh38"</code>). If set <code>NULL</code>, defaults to the most recent build available.
Note: don't pass in UCSC build IDs (e.g. <code>"hg38"</code>).</p></td>
    </tr>
    <tr>
      <th>gffFile</th>
      <td><p><code>string</code>. GFF/GTF (General Feature Format) file.
Generally, we recommend using a GTF (GFFv2) instead of a GFFv3 file if
possible.</p></td>
    </tr>
    <tr>
      <th>transgeneNames</th>
      <td><p><code>character</code>. Vector indicating which assay
rows denote transgenes (e.g. EGFP, TDTOMATO).</p></td>
    </tr>
    <tr>
      <th>spikeNames</th>
      <td><p><code>character</code>. Vector indicating which assay rows denote
spike-in sequences (e.g. ERCCs).</p></td>
    </tr>
    <tr>
      <th>interestingGroups</th>
      <td><p><code>character</code>. Groups of interest that
define the samples. If left unset, defaults to <code>sampleName</code>.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p><code>bcbioSingleCell</code>.</p>
    
    <h2 class="hasAnchor" id="remote-data"><a class="anchor" href="#remote-data"></a>Remote Data</h2>

    
    <p>When working in RStudio, we recommend connecting to the bcbio-nextgen run
directory as a remote connection over
<a href='https://github.com/osxfuse/osxfuse/wiki/SSHFS'>sshfs</a>.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><ul>
<li><p><code><a href='https://www.rdocumentation.org/packages/SingleCellExperiment/topics/SingleCellExperiment'>SingleCellExperiment::SingleCellExperiment()</a></code>.</p></li>
<li><p><code>.S4methods(class = "bcbioSingleCell")</code>.</p></li>
</ul></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>uploadDir</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/system.file'>system.file</a></span>(<span class='st'>"extdata/indrops"</span>, <span class='kw'>package</span> <span class='kw'>=</span> <span class='st'>"bcbioSingleCell"</span>)

<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>bcbioSingleCell</span>(<span class='no'>uploadDir</span>)</div><div class='output co'>#&gt; <span class='message'>Dated project directory: 2018-01-01_bcbio</span></div><div class='output co'>#&gt; <span class='message'>1 sample(s) detected:</span>
#&gt; <span class='message'>multiplexed_AAAAAAAA</span></div><div class='output co'>#&gt; <span class='message'>Importing project-summary.yaml using yaml::yaml.load_file().</span></div><div class='output co'>#&gt; <span class='message'>Data versions are missing.</span></div><div class='output co'>#&gt; <span class='message'>Importing bcbio-nextgen.log using readr::read_lines().</span></div><div class='output co'>#&gt; <span class='message'>Importing bcbio-nextgen-commands.log using readr::read_lines().</span></div><div class='output co'>#&gt; <span class='message'>1000 reads per cellular barcode cutoff detected.</span></div><div class='output co'>#&gt; <span class='message'>Counts will imported as genes.</span></div><div class='output co'>#&gt; <span class='message'>UMI type: harvard-indrop-v3</span></div><div class='output co'>#&gt; <span class='message'>Importing unfiltered cellular barcode distributions.</span></div><div class='output co'>#&gt; <span class='message'>Importing counts.</span></div><div class='output co'>#&gt; <span class='message'>Imported multiplexed-AAAAAAAA.</span></div><div class='output co'>#&gt; <span class='message'>Unknown organism. Skipping annotations.</span></div><div class='output co'>#&gt; <span class='message'>`sampleMetadataFile` is recommended for multiplexed samples (e.g. harvard-indrop-v3).</span></div><div class='output co'>#&gt; <span class='message'>Calculating cellular barcode metrics.</span></div><div class='output co'>#&gt; <span class='message'>100 cells detected.</span></div><div class='output co'>#&gt; <span class='message'>Calculating metrics without biotype information.</span>
#&gt; <span class='message'>`rowRanges` is required to calculate: nCoding, nMito, mitoRatio</span></div><div class='output co'>#&gt; <span class='message'>100 / 100 cellular barcodes passed pre-filtering (100%)</span></div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/print'>print</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; bcbioSingleCell 0.3.4
#&gt; uploadDir: /Users/mike/git/bcbioSingleCell/inst/extdata/indrops
#&gt; dates(2): [bcbio] 2018-01-01; [R] 2018-12-01
#&gt; level: genes
#&gt; interestingGroups: sampleName
#&gt; filtered: FALSE
#&gt; class: SingleCellExperiment 
#&gt; dim: 50 100 
#&gt; metadata(27): version pipeline ... wd sessionInfo
#&gt; assays(1): counts
#&gt; rownames(50): ENSG00000071082 ENSG00000100316 ... ENSG00000269028
#&gt;   ENSG00000282105
#&gt; rowData names(0):
#&gt; colnames(100): AAACACTA_CTTCGATT AAACTACA_CCACATTA ...
#&gt;   TGGGAATT_ATATAGGA TGTTATCA_ACGCAGAG
#&gt; colData names(9): nUMI nGene ... sampleID sampleName
#&gt; reducedDimNames(0):
#&gt; spikeNames(0):</div><div class='input'>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>bcbioSingleCell</span>(
    <span class='kw'>uploadDir</span> <span class='kw'>=</span> <span class='no'>uploadDir</span>,
    <span class='kw'>sampleMetadataFile</span> <span class='kw'>=</span> <span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/file.path'>file.path</a></span>(<span class='no'>uploadDir</span>, <span class='st'>"metadata.csv"</span>)
)</div><div class='output co'>#&gt; <span class='message'>Dated project directory: 2018-01-01_bcbio</span></div><div class='output co'>#&gt; <span class='message'>1 sample(s) detected:</span>
#&gt; <span class='message'>multiplexed_AAAAAAAA</span></div><div class='output co'>#&gt; <span class='message'>Importing project-summary.yaml using yaml::yaml.load_file().</span></div><div class='output co'>#&gt; <span class='message'>Data versions are missing.</span></div><div class='output co'>#&gt; <span class='message'>Importing bcbio-nextgen.log using readr::read_lines().</span></div><div class='output co'>#&gt; <span class='message'>Importing bcbio-nextgen-commands.log using readr::read_lines().</span></div><div class='output co'>#&gt; <span class='message'>1000 reads per cellular barcode cutoff detected.</span></div><div class='output co'>#&gt; <span class='message'>Counts will imported as genes.</span></div><div class='output co'>#&gt; <span class='message'>UMI type: harvard-indrop-v3</span></div><div class='output co'>#&gt; <span class='message'>Importing metadata.csv using data.table::fread().</span></div><div class='output co'>#&gt; <span class='message'>Multiplexed samples detected.</span></div><div class='output co'>#&gt; <span class='message'>Importing unfiltered cellular barcode distributions.</span></div><div class='output co'>#&gt; <span class='message'>Importing counts.</span></div><div class='output co'>#&gt; <span class='message'>Imported multiplexed-AAAAAAAA.</span></div><div class='output co'>#&gt; <span class='message'>Unknown organism. Skipping annotations.</span></div><div class='output co'>#&gt; <span class='message'>Calculating cellular barcode metrics.</span></div><div class='output co'>#&gt; <span class='message'>100 cells detected.</span></div><div class='output co'>#&gt; <span class='message'>Calculating metrics without biotype information.</span>
#&gt; <span class='message'>`rowRanges` is required to calculate: nCoding, nMito, mitoRatio</span></div><div class='output co'>#&gt; <span class='message'>100 / 100 cellular barcodes passed pre-filtering (100%)</span></div><div class='input'><span class='fu'><a href='https://www.rdocumentation.org/packages/base/topics/print'>print</a></span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; bcbioSingleCell 0.3.4
#&gt; uploadDir: /Users/mike/git/bcbioSingleCell/inst/extdata/indrops
#&gt; dates(2): [bcbio] 2018-01-01; [R] 2018-12-01
#&gt; level: genes
#&gt; sampleMetadataFile: /Users/mike/git/bcbioSingleCell/inst/extdata/indrops/metadata.csv
#&gt; interestingGroups: sampleName
#&gt; filtered: FALSE
#&gt; class: SingleCellExperiment 
#&gt; dim: 50 100 
#&gt; metadata(27): version pipeline ... wd sessionInfo
#&gt; assays(1): counts
#&gt; rownames(50): ENSG00000071082 ENSG00000100316 ... ENSG00000269028
#&gt;   ENSG00000282105
#&gt; rowData names(0):
#&gt; colnames(100): AAACACTA_CTTCGATT AAACTACA_CCACATTA ...
#&gt;   TGGGAATT_ATATAGGA TGTTATCA_ACGCAGAG
#&gt; colData names(15): nUMI nGene ... aggregate revcomp
#&gt; reducedDimNames(0):
#&gt; spikeNames(0):</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#remote-data">Remote Data</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    <p>Michael Steinbaugh, Rory Kirchner</p>
  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com/'>Michael Steinbaugh</a>, <a href='https://github.com/roryk/'>Rory Kirchner</a>, <a href='https://github.com/marypiper/'>Mary Piper</a>, <a href='https://github.com/vbarrera/'>Victor Barrera</a>, <a href='https://github.com/sjhosui/'>Shannan Ho Sui</a>, <a href='https://bioinformatics.sph.harvard.edu/'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.2.0.</p>
</div>
      </footer>
   </div>

  

  </body>
</html>

