<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Differential Expression — diffExp • bcbioSingleCell</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Differential Expression — diffExp" />

<meta property="og:description" content="We now generally recommend the ZINB-WaVE method over zingeR, since it is
faster, and has been show to be more sensitive for most single-cell RNA-seq
datasets." />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioSingleCell</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.1.7</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioSingleCell">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Differential Expression</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioSingleCell/blob/master/R/AllGenerics.R'><code>R/AllGenerics.R</code></a>, <a href='https://github.com/hbc/bcbioSingleCell/blob/master/R/methods-diffExp.R'><code>R/methods-diffExp.R</code></a></small>
    <div class="hidden name"><code>diffExp.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>We now generally recommend the ZINB-WaVE method over zingeR, since it is
faster, and has been show to be more sensitive for most single-cell RNA-seq
datasets.</p>
    
    </div>

    <pre class="usage"><span class='fu'>diffExp</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for SingleCellExperiment</span>
<span class='fu'>diffExp</span>(<span class='no'>object</span>, <span class='no'>numerator</span>, <span class='no'>denominator</span>,
  <span class='kw'>zeroWeights</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"zinbwave"</span>, <span class='st'>"zingeR"</span>), <span class='kw'>caller</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"edgeR"</span>, <span class='st'>"DESeq2"</span>),
  <span class='kw'>minCellsPerGene</span> <span class='kw'>=</span> <span class='fl'>25L</span>, <span class='kw'>minCountsPerCell</span> <span class='kw'>=</span> <span class='fl'>5L</span>)

<span class='co'># S4 method for seurat</span>
<span class='fu'>diffExp</span>(<span class='no'>object</span>, <span class='no'>numerator</span>, <span class='no'>denominator</span>,
  <span class='kw'>zeroWeights</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"zinbwave"</span>, <span class='st'>"zingeR"</span>), <span class='kw'>caller</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"edgeR"</span>, <span class='st'>"DESeq2"</span>),
  <span class='kw'>minCellsPerGene</span> <span class='kw'>=</span> <span class='fl'>25L</span>, <span class='kw'>minCountsPerCell</span> <span class='kw'>=</span> <span class='fl'>5L</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    <tr>
      <th>numerator</th>
      <td><p>Group of cells to use in the numerator of the contrast
(e.g. treatment).</p></td>
    </tr>
    <tr>
      <th>denominator</th>
      <td><p>Group of cells to use in the denominator of the contrast
(e.g. control).</p></td>
    </tr>
    <tr>
      <th>zeroWeights</th>
      <td><p>Package to use for zero weight calculations. Defaults to
zinbwave but zingeR is also supported.</p></td>
    </tr>
    <tr>
      <th>caller</th>
      <td><p>Package to use for differential expression calling. Defaults to
edgeR (faster for large datasets) but DESeq2 is also supported.</p></td>
    </tr>
    <tr>
      <th>minCellsPerGene</th>
      <td><p>The minimum number of cells where a gene is expressed,
to pass low expression filtering. Set to <code>0</code> to disable (not recommended).</p></td>
    </tr>
    <tr>
      <th>minCountsPerCell</th>
      <td><p>Minimum number of counts per cell for a gene to pass
low expression filtering. The number of cells is defined by
<code>minCellsPerGene</code>. Set to <code>0</code> to disable (not recommended).</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    
<ul>
<li><p><code>caller = "edgeR"</code>: <code>DEGLRT</code>.</p></li>
<li><p><code>caller = "DESeq2"</code>: Unshrunken <code>DESeqResults</code>. Use <code>lfcShrink()</code> if
shrunken results are desired.</p></li>
</ul>

    
    <h2 class="hasAnchor" id="zinbwave"><a class="anchor" href="#zinbwave"></a>zinbwave</h2>

    
    <p>We are currently using an epsilon setting of 1e12, as recommended by the
ZINB-WaVE integration paper.</p>
    
    <h2 class="hasAnchor" id="zinger"><a class="anchor" href="#zinger"></a>zingeR</h2>

    
    <p>We first perform a differential expression analysis using zingeR posterior
probabilities and focussing on the count component of the ZINB model. The
weights are estimated with the core function of zingeR, <code>zeroWeightsLS()</code>. It
is important to be consistent with the normalization procedure, i.e. if you
use TMM normalization for the analysis, you should also use it for estimating
the zingeR posterior probabilities.</p>
    
    <h2 class="hasAnchor" id="edger"><a class="anchor" href="#edger"></a>edgeR</h2>

    
    <p>After estimation of the dispersions and posterior probabilities, the
<code>glmWeightedF()</code> function is used for statistical inference. This is an
adapted function from the <code>glmLRT()</code> function of edgeR. It uses an F-test for
which the denominator degrees of freedom are by default adjusted according to
the downweighting of excess zeros (<code>ZI = TRUE</code>). Also, independent filtering
can be performed on the obtained p-values (<code>independentFiltering = TRUE</code>). We
use the independent filtering strategy that was originally implemented in
DESeq2. By default, the average fitted values are used as a filter criterion.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Consult the following sources for more information:</p><ul>
<li><p><a href='https://doi.org/10.1186/s13059-018-1406-4'>Observation weights unlock bulk RNA-seq tools for zero inflation and single-cell applications</a>.</p></li>
<li><p><a href='https://goo.gl/4rTK1w'>zingeR vignette</a>.</p></li>
<li><p><a href='https://bit.ly/2wtDdpS'>zinbwave vignette</a>.</p></li>
<li><p><a href='https://github.com/mikelove/zinbwave-deseq2'>zimbwave-DESeq2 workflow</a>.</p></li>
</ul></div>
    <div class='dont-index'><ul>
<li><p>DESeq2: We're trying to follow the conventions used in DESeq2 for
contrasts, defining the name of the factor in the design formula,
numerator, and denominator level for the fold change calculations. See
<code><a href='http://www.rdocumentation.org/packages/DESeq2/topics/results'>DESeq2::results()</a></code> for more information.</p></li>
<li><p>Seurat: Note that Seurat currently uses the convention <code>cells.1</code> for the
numerator and <code>cells.2</code> for the denominator. See <code><a href='http://www.rdocumentation.org/packages/Seurat/topics/DiffExpTest'>Seurat::DiffExpTest()</a></code>
for additional information.</p></li>
</ul></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='co'># SingleCellExperiment ====</span>
<span class='no'>m</span> <span class='kw'>&lt;-</span> <span class='fu'><a href='metrics.html'>metrics</a></span>(<span class='no'>cellranger_small</span>)
<span class='no'>numerator</span> <span class='kw'>&lt;-</span> <span class='fu'>rownames</span>(<span class='no'>m</span>)[<span class='fu'>which</span>(<span class='no'>m</span><span class='kw'>[[</span><span class='st'>"sampleName"</span>]] <span class='kw'>==</span> <span class='st'>"proximal"</span>)]
<span class='no'>denominator</span> <span class='kw'>&lt;-</span> <span class='fu'>rownames</span>(<span class='no'>m</span>)[<span class='fu'>which</span>(<span class='no'>m</span><span class='kw'>[[</span><span class='st'>"sampleName"</span>]] <span class='kw'>==</span> <span class='st'>"distal"</span>)]

<span class='co'># zinbwave-DESeq2</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>diffExp</span>(
    <span class='kw'>object</span> <span class='kw'>=</span> <span class='no'>cellranger_small</span>,
    <span class='kw'>numerator</span> <span class='kw'>=</span> <span class='no'>numerator</span>,
    <span class='kw'>denominator</span> <span class='kw'>=</span> <span class='no'>denominator</span>,
    <span class='kw'>zeroWeights</span> <span class='kw'>=</span> <span class='st'>"zinbwave"</span>,
    <span class='kw'>caller</span> <span class='kw'>=</span> <span class='st'>"DESeq2"</span>
)</div><div class='output co'>#&gt; <span class='message'>Loading required namespace: zinbwave</span></div><div class='output co'>#&gt; <span class='message'>Loading required namespace: DESeq2</span></div><div class='output co'>#&gt; <span class='message'>Performing differential expression with zinbwave-DESeq2</span></div><div class='output co'>#&gt; <span class='message'>Applying low gene count filter</span>
#&gt; <span class='message'>Requiring at least 25 cells with counts of 5 or more</span></div><div class='output co'>#&gt; genes
#&gt; TRUE 
#&gt;  500 
#&gt;    user  system elapsed 
#&gt; 102.367   2.503  35.256 </div><div class='output co'>#&gt; <span class='message'>converting counts to integer mode</span></div><div class='output co'>#&gt; <span class='message'>estimating size factors</span></div><div class='output co'>#&gt; <span class='message'>estimating dispersions</span></div><div class='output co'>#&gt; <span class='message'>gene-wise dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>mean-dispersion relationship</span></div><div class='output co'>#&gt; <span class='message'>final dispersion estimates</span></div><div class='output co'>#&gt; <span class='message'>fitting model and testing</span></div><div class='output co'>#&gt;    user  system elapsed 
#&gt; 176.828   2.868 182.050 </div><div class='img'><img src='diffExp-1.png' alt='' width='700' height='433' /></div><div class='output co'>#&gt; $rect
#&gt; $rect$w
#&gt; [1] 1.188434
#&gt; 
#&gt; $rect$h
#&gt; [1] 0.35884
#&gt; 
#&gt; $rect$left
#&gt; [1] 1.909997
#&gt; 
#&gt; $rect$top
#&gt; [1] -1.358794
#&gt; 
#&gt; 
#&gt; $text
#&gt; $text$x
#&gt; [1] 2.258307 2.258307 2.258307
#&gt; 
#&gt; $text$y
#&gt; [1] -1.448504 -1.538214 -1.627924
#&gt; 
#&gt; </div><div class='input'><span class='fu'>class</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "DESeqResults"
#&gt; attr(,"package")
#&gt; [1] "DESeq2"</div><div class='input'><span class='fu'>head</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; log2 fold change (MLE): group numerator vs denominator 
#&gt; LRT p-value: '~ group' vs '~ 1' 
#&gt; DataFrame with 6 rows and 6 columns
#&gt;                         baseMean     log2FoldChange              lfcSE             stat               pvalue
#&gt;                        &lt;numeric&gt;          &lt;numeric&gt;          &lt;numeric&gt;        &lt;numeric&gt;            &lt;numeric&gt;
#&gt; ENSG00000005022 40.6946883919453 0.0969614143415067 0.0583898490050989 2.74885559613085   0.0973240650417439
#&gt; ENSG00000008018 14.8932881299383  0.592146719221284 0.0635287124600374 86.8346534696238 1.17984003724583e-20
#&gt; ENSG00000008394 39.5480029505622   1.18187261290727 0.0911514348619983 152.376927876437 5.24138622879893e-35
#&gt; ENSG00000008988 217.581034338649 -0.635753413462794 0.0360206306430627 325.539996396201 8.99915778524178e-73
#&gt; ENSG00000026025 18.0728087783362  -0.68221630287468  0.242477684485772 7.98031541354476  0.00472887021635307
#&gt; ENSG00000029993 13.9244947453213   -1.2302100247238  0.145219274973856 78.0883321336796  9.8534261071521e-19
#&gt;                                 padj
#&gt;                            &lt;numeric&gt;
#&gt; ENSG00000005022    0.108190905285977
#&gt; ENSG00000008018 2.77905449531364e-20
#&gt; ENSG00000008394 1.91541834978902e-34
#&gt; ENSG00000008988 1.27788040550433e-71
#&gt; ENSG00000026025  0.00577456633299134
#&gt; ENSG00000029993  2.1765123445576e-18</div><div class='input'>
<span class='co'># zinbwave-edgeR</span>
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>diffExp</span>(
    <span class='kw'>object</span> <span class='kw'>=</span> <span class='no'>cellranger_small</span>,
    <span class='kw'>numerator</span> <span class='kw'>=</span> <span class='no'>numerator</span>,
    <span class='kw'>denominator</span> <span class='kw'>=</span> <span class='no'>denominator</span>,
    <span class='kw'>zeroWeights</span> <span class='kw'>=</span> <span class='st'>"zinbwave"</span>,
    <span class='kw'>caller</span> <span class='kw'>=</span> <span class='st'>"edgeR"</span>
)</div><div class='output co'>#&gt; <span class='message'>Performing differential expression with zinbwave-edgeR</span></div><div class='output co'>#&gt; <span class='message'>Applying low gene count filter</span>
#&gt; <span class='message'>Requiring at least 25 cells with counts of 5 or more</span></div><div class='output co'>#&gt; genes
#&gt; TRUE 
#&gt;  500 
#&gt;    user  system elapsed 
#&gt;  30.777   0.946  32.260 </div><div class='input'><span class='fu'>class</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "DGELRT"
#&gt; attr(,"package")
#&gt; [1] "edgeR"</div><div class='input'><span class='fu'>head</span>(<span class='no'>x</span><span class='kw'>[[</span><span class='st'>"table"</span>]])</div><div class='output co'>#&gt;                      logFC    logCPM         LR       PValue
#&gt; ENSG00000005022  0.1774111 10.050237   9.141339 2.626271e-03
#&gt; ENSG00000008018  0.6793362  8.731102 106.773456 7.863967e-23
#&gt; ENSG00000008394  1.2591768 10.027434 168.536534 1.887592e-33
#&gt; ENSG00000008988 -0.5308476 12.389327 308.188318 3.441931e-54
#&gt; ENSG00000026025 -0.5718466  9.516885   5.725923 1.729285e-02
#&gt; ENSG00000029993 -1.1655945  8.681385  69.050417 9.760156e-16</div><div class='input'>
<span class='co'># seurat ====</span>
<span class='co'># Expression in cluster 3 relative to cluster 2</span>
<span class='no'>numerator</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>Seurat</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/Seurat/topics/WhichCells'>WhichCells</a></span>(<span class='kw pkg'>Seurat</span><span class='kw ns'>::</span><span class='no'><a href='http://www.rdocumentation.org/packages/Seurat/topics/pbmc_small'>pbmc_small</a></span>, <span class='kw'>ident</span> <span class='kw'>=</span> <span class='fl'>3L</span>)
<span class='no'>denominator</span> <span class='kw'>&lt;-</span> <span class='kw pkg'>Seurat</span><span class='kw ns'>::</span><span class='fu'><a href='http://www.rdocumentation.org/packages/Seurat/topics/WhichCells'>WhichCells</a></span>(<span class='kw pkg'>Seurat</span><span class='kw ns'>::</span><span class='no'><a href='http://www.rdocumentation.org/packages/Seurat/topics/pbmc_small'>pbmc_small</a></span>, <span class='kw'>ident</span> <span class='kw'>=</span> <span class='fl'>2L</span>)
<span class='no'>x</span> <span class='kw'>&lt;-</span> <span class='fu'>diffExp</span>(
    <span class='kw'>object</span> <span class='kw'>=</span> <span class='kw pkg'>Seurat</span><span class='kw ns'>::</span><span class='no'><a href='http://www.rdocumentation.org/packages/Seurat/topics/pbmc_small'>pbmc_small</a></span>,
    <span class='kw'>numerator</span> <span class='kw'>=</span> <span class='no'>numerator</span>,
    <span class='kw'>denominator</span> <span class='kw'>=</span> <span class='no'>denominator</span>,
    <span class='kw'>caller</span> <span class='kw'>=</span> <span class='st'>"edgeR"</span>,
    <span class='kw'>minCellsPerGene</span> <span class='kw'>=</span> <span class='fl'>5L</span>,
    <span class='kw'>minCountsPerCell</span> <span class='kw'>=</span> <span class='fl'>5L</span>
)</div><div class='output co'>#&gt; <span class='message'>Performing differential expression with zinbwave-edgeR</span></div><div class='output co'>#&gt; <span class='message'>Applying low gene count filter</span>
#&gt; <span class='message'>Requiring at least 5 cells with counts of 5 or more</span></div><div class='output co'>#&gt; genes
#&gt; FALSE  TRUE 
#&gt;   213    17 
#&gt;    user  system elapsed 
#&gt;   0.929   0.015   0.947 </div><div class='input'><span class='fu'>class</span>(<span class='no'>x</span>)</div><div class='output co'>#&gt; [1] "DGELRT"
#&gt; attr(,"package")
#&gt; [1] "edgeR"</div><div class='input'><span class='fu'>head</span>(<span class='no'>x</span><span class='kw'>[[</span><span class='st'>"table"</span>]])</div><div class='output co'>#&gt;             logFC   logCPM           LR      PValue
#&gt; NKG7   -8.0978765 17.95711 4.001607e-05 0.995039029
#&gt; CCL5    0.3327462 16.86562 1.538144e+00 0.228265805
#&gt; TYROBP  3.6536228 16.70005 1.240148e+01 0.003558917
#&gt; LGALS1  1.8190557 16.37732 1.723488e+00 0.207087609
#&gt; FCER1G  0.1598044 15.97882 1.737223e-02 0.896824369
#&gt; PPBP    9.3502205 19.18038 1.125326e-08 0.999917035</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#zinbwave">zinbwave</a></li>

      <li><a href="#zinger">zingeR</a></li>

      <li><a href="#edger">edgeR</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

