<!-- Generated by pkgdown: do not edit by hand -->
<!DOCTYPE html>
<html>
  <head>
  <meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">

<title>Plot Cell-Type-Specific Gene Markers — plotKnownMarkersDetected • bcbioSingleCell</title>

<!-- jquery -->
<script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script>
<!-- Bootstrap -->

<link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>

<!-- Font Awesome icons -->
<link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">

<!-- clipboard.js -->
<script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/1.7.1/clipboard.min.js" integrity="sha384-cV+rhyOuRHc9Ub/91rihWcGmMmCXDeksTtCihMupQHSsi8GIIRDG0ThDc3HGQFJ3" crossorigin="anonymous"></script>

<!-- pkgdown -->
<link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script>
<script src="../pkgdown.js"></script>



  <link href="../extra.css" rel="stylesheet">
  
<meta property="og:title" content="Plot Cell-Type-Specific Gene Markers — plotKnownMarkersDetected" />

<meta property="og:description" content="Plot gene expression per cell in multiple formats:
plotMarkerTSNE(): t-SNE gene expression plot.
plotDot(): Dot plot.
plotViolin(): Violin plot.
" />
<meta name="twitter:card" content="summary" />


<!-- mathjax -->
<script src='https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML'></script>

<!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->


  </head>

  <body>
    <div class="container template-reference-topic">
      <header>
      <div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">bcbioSingleCell</a>
        <span class="label label-default" data-toggle="tooltip" data-placement="bottom" title="Released package">0.1.11</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">Changelog</a>
</li>
      </ul>
      
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/hbc/bcbioSingleCell">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
      
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      
      </header>

<div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Plot Cell-Type-Specific Gene Markers</h1>
    <small class="dont-index">Source: <a href='https://github.com/hbc/bcbioSingleCell/blob/master/R/AllGenerics.R'><code>R/AllGenerics.R</code></a>, <a href='https://github.com/hbc/bcbioSingleCell/blob/master/R/methods-plotMarker.R'><code>R/methods-plotMarker.R</code></a></small>
    <div class="hidden name"><code>plotMarker.Rd</code></div>
    </div>

    <div class="ref-description">
    
    <p>Plot gene expression per cell in multiple formats:</p><ol>
<li><p><code>plotMarkerTSNE()</code>: t-SNE gene expression plot.</p></li>
<li><p><code><a href='plotDot.html'>plotDot()</a></code>: Dot plot.</p></li>
<li><p><code><a href='plotViolin.html'>plotViolin()</a></code>: Violin plot.</p></li>
</ol>
    
    </div>

    <pre class="usage"><span class='fu'>plotKnownMarkersDetected</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='fu'>plotMarkerTSNE</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='fu'>plotMarkerUMAP</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='fu'>plotTopMarkers</span>(<span class='no'>object</span>, <span class='no'>...</span>)

<span class='co'># S4 method for seurat</span>
<span class='fu'>plotMarkerTSNE</span>(<span class='no'>object</span>, <span class='no'>genes</span>, <span class='kw'>expression</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"mean"</span>,
  <span class='st'>"median"</span>, <span class='st'>"sum"</span>), <span class='kw'>color</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>pointsAsNumbers</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>pointSize</span> <span class='kw'>=</span> <span class='fl'>0.75</span>, <span class='kw'>pointAlpha</span> <span class='kw'>=</span> <span class='fl'>0.8</span>, <span class='kw'>label</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>labelSize</span> <span class='kw'>=</span> <span class='fl'>6L</span>,
  <span class='kw'>dark</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>grid</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>legend</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>aspectRatio</span> <span class='kw'>=</span> <span class='fl'>1L</span>,
  <span class='kw'>title</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='co'># S4 method for seurat</span>
<span class='fu'>plotMarkerUMAP</span>(<span class='no'>object</span>, <span class='no'>genes</span>, <span class='kw'>expression</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"mean"</span>,
  <span class='st'>"median"</span>, <span class='st'>"sum"</span>), <span class='kw'>color</span> <span class='kw'>=</span> <span class='kw'>NULL</span>, <span class='kw'>pointsAsNumbers</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
  <span class='kw'>pointSize</span> <span class='kw'>=</span> <span class='fl'>0.75</span>, <span class='kw'>pointAlpha</span> <span class='kw'>=</span> <span class='fl'>0.8</span>, <span class='kw'>label</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>labelSize</span> <span class='kw'>=</span> <span class='fl'>6L</span>,
  <span class='kw'>dark</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>grid</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>, <span class='kw'>legend</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>, <span class='kw'>aspectRatio</span> <span class='kw'>=</span> <span class='fl'>1L</span>,
  <span class='kw'>title</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)

<span class='co'># S4 method for seurat</span>
<span class='fu'>plotTopMarkers</span>(<span class='no'>object</span>, <span class='no'>markers</span>, <span class='kw'>dimRed</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"tsne"</span>, <span class='st'>"umap"</span>),
  <span class='kw'>headerLevel</span> <span class='kw'>=</span> <span class='fl'>2L</span>, <span class='no'>...</span>)

<span class='co'># S4 method for seurat</span>
<span class='fu'>plotKnownMarkersDetected</span>(<span class='no'>object</span>, <span class='no'>markers</span>,
  <span class='kw'>dimRed</span> <span class='kw'>=</span> <span class='fu'>c</span>(<span class='st'>"tsne"</span>, <span class='st'>"umap"</span>), <span class='kw'>headerLevel</span> <span class='kw'>=</span> <span class='fl'>2L</span>, <span class='no'>...</span>)</pre>
    
    <h2 class="hasAnchor" id="arguments"><a class="anchor" href="#arguments"></a>Arguments</h2>
    <table class="ref-arguments">
    <colgroup><col class="name" /><col class="desc" /></colgroup>
    <tr>
      <th>object</th>
      <td><p>Object.</p></td>
    </tr>
    <tr>
      <th>...</th>
      <td><p>Additional arguments.</p></td>
    </tr>
    <tr>
      <th>genes</th>
      <td><p>Gene identifiers. Must match the rownames of the object.</p></td>
    </tr>
    <tr>
      <th>expression</th>
      <td><p>Calculation to apply. Uses <code>match.arg()</code> and defaults to
the first argument in the vector.</p></td>
    </tr>
    <tr>
      <th>color</th>
      <td><p>Desired ggplot color scale. Must supply discrete values. When
set to <code>NULL</code>, the default ggplot2 color palette will be used. If manual
color definitions are desired, we recommend using
<code><a href='http://www.rdocumentation.org/packages/ggplot2/topics/scale_manual'>ggplot2::scale_color_manual()</a></code>.</p></td>
    </tr>
    <tr>
      <th>pointsAsNumbers</th>
      <td><p>Plot the points as numbers (<code>TRUE</code>) or dots (<code>FALSE</code>).</p></td>
    </tr>
    <tr>
      <th>pointSize</th>
      <td><p>Cell point size.</p></td>
    </tr>
    <tr>
      <th>pointAlpha</th>
      <td><p>Alpha transparency level. Useful when there many cells in
the dataset, and some cells can be masked.</p></td>
    </tr>
    <tr>
      <th>label</th>
      <td><p>Overlay a cluster identitiy label on the plot.</p></td>
    </tr>
    <tr>
      <th>labelSize</th>
      <td><p>Size of the text label.</p></td>
    </tr>
    <tr>
      <th>dark</th>
      <td><p>Plot against a dark background using
<code><a href='http://www.rdocumentation.org/packages/basejump/topics/themes'>basejump::theme_midnight()</a></code>.</p></td>
    </tr>
    <tr>
      <th>grid</th>
      <td><p>Show major grid lines but hide axis lines.</p></td>
    </tr>
    <tr>
      <th>legend</th>
      <td><p>Include plot legend.</p></td>
    </tr>
    <tr>
      <th>aspectRatio</th>
      <td><p>Aspect ratio.</p></td>
    </tr>
    <tr>
      <th>title</th>
      <td><p>Plot title.</p></td>
    </tr>
    <tr>
      <th>markers</th>
      <td><p><code>grouped_df</code> of marker genes.</p><ul>
<li><p><code>plotTopMarkers()</code>: must be grouped by "<code>cluster</code>".</p></li>
<li><p><code>plotKnownMarkersDetected()</code>: must be grouped by "<code>cellType</code>".</p></li>
</ul></td>
    </tr>
    <tr>
      <th>dimRed</th>
      <td><p>Dimensionality reduction method to apply. Defaults to t-SNE
("<code>tsne</code>") but UMAP is also supported ("<code>umap</code>").</p></td>
    </tr>
    <tr>
      <th>headerLevel</th>
      <td><p>R Markdown header level.</p></td>
    </tr>
    </table>
    
    <h2 class="hasAnchor" id="value"><a class="anchor" href="#value"></a>Value</h2>

    <p>Show graphical output. Invisibly return <code>ggplot</code> <code>list</code>.</p>
    
    <h2 class="hasAnchor" id="plottopmarkers"><a class="anchor" href="#plottopmarkers"></a>plotTopMarkers</h2>

    
    <p>The number of markers to plot is determined by the output of the
<code>topMarkers()</code> function. If you want to reduce the number of genes to plot,
simply reassign first using that function. If necessary, we can add support
for the number of genes to plot here in a future update.</p>
    
    <h2 class="hasAnchor" id="see-also"><a class="anchor" href="#see-also"></a>See also</h2>

    <div class='dont-index'><p>Other Clustering Functions: <code>cellTypesPerCluster</code>,
  <code>knownMarkersDetected</code>,
  <code>plotCellTypesPerCluster</code>,
  <code><a href='plotFeature.html'>plotFeatureTSNE</a></code>, <code>plotPCElbow</code>,
  <code><a href='plotDimensionalReduction.html'>plotTSNE</a></code>, <code>sanitizeMarkers</code>,
  <code>topMarkers</code></p></div>
    

    <h2 class="hasAnchor" id="examples"><a class="anchor" href="#examples"></a>Examples</h2>
    <pre class="examples"><div class='input'><span class='no'>object</span> <span class='kw'>&lt;-</span> <span class='no'>seurat_small</span>
<span class='no'>collagen_genes</span> <span class='kw'>&lt;-</span> <span class='fu'>grep</span>(<span class='st'>"^COL\\d"</span>, <span class='fu'>rownames</span>(<span class='no'>object</span>), <span class='kw'>value</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='fu'>print</span>(<span class='no'>collagen_genes</span>)</div><div class='output co'>#&gt;  [1] "COL1A1"  "COL5A1"  "COL4A2"  "COL6A1"  "COL6A2"  "COL6A3"  "COL1A2" 
#&gt;  [8] "COL3A1"  "COL18A1" "COL5A2" </div><div class='input'><span class='no'>mito_genes</span> <span class='kw'>&lt;-</span> <span class='fu'>grep</span>(<span class='st'>"^MT-"</span>, <span class='fu'>rownames</span>(<span class='no'>object</span>), <span class='kw'>value</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>)
<span class='fu'>print</span>(<span class='no'>mito_genes</span>)</div><div class='output co'>#&gt;  [1] "MT-ND6"  "MT-CO2"  "MT-CYB"  "MT-ND2"  "MT-ND5"  "MT-CO1"  "MT-ND3" 
#&gt;  [8] "MT-ND4"  "MT-ND1"  "MT-ATP6" "MT-CO3"  "MT-ND4L" "MT-ATP8"</div><div class='input'>
<span class='co'># t-SNE</span>
<span class='fu'>plotMarkerTSNE</span>(
    <span class='kw'>object</span> <span class='kw'>=</span> <span class='no'>object</span>,
    <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>collagen_genes</span>,
    <span class='kw'>dark</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
    <span class='kw'>grid</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
    <span class='kw'>title</span> <span class='kw'>=</span> <span class='st'>"collagen"</span>
)</div><div class='img'><img src='plotMarker-1.png' alt='' width='700' height='433' /></div><div class='input'><span class='fu'>plotMarkerTSNE</span>(
    <span class='kw'>object</span> <span class='kw'>=</span> <span class='no'>object</span>,
    <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>mito_genes</span>,
    <span class='kw'>dark</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
    <span class='kw'>grid</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
    <span class='kw'>title</span> <span class='kw'>=</span> <span class='st'>"mito"</span>
)</div><div class='img'><img src='plotMarker-2.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Number cloud</span>
<span class='fu'>plotMarkerTSNE</span>(
    <span class='kw'>object</span> <span class='kw'>=</span> <span class='no'>object</span>,
    <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>collagen_genes</span>,
    <span class='kw'>pointsAsNumbers</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>
)</div><div class='img'><img src='plotMarker-3.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># UMAP</span>
<span class='fu'>plotMarkerUMAP</span>(
    <span class='kw'>object</span> <span class='kw'>=</span> <span class='no'>object</span>,
    <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>collagen_genes</span>,
    <span class='kw'>dark</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
    <span class='kw'>grid</span> <span class='kw'>=</span> <span class='fl'>TRUE</span>,
    <span class='kw'>title</span> <span class='kw'>=</span> <span class='st'>"collagen"</span>
)</div><div class='img'><img src='plotMarker-4.png' alt='' width='700' height='433' /></div><div class='input'><span class='fu'>plotMarkerUMAP</span>(
    <span class='kw'>object</span> <span class='kw'>=</span> <span class='no'>object</span>,
    <span class='kw'>genes</span> <span class='kw'>=</span> <span class='no'>mito_genes</span>,
    <span class='kw'>dark</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
    <span class='kw'>grid</span> <span class='kw'>=</span> <span class='fl'>FALSE</span>,
    <span class='kw'>title</span> <span class='kw'>=</span> <span class='st'>"mito"</span>
)</div><div class='img'><img src='plotMarker-5.png' alt='' width='700' height='433' /></div><div class='input'>
<span class='co'># Top markers</span>
<span class='no'>markers</span> <span class='kw'>&lt;-</span> <span class='fu'>topMarkers</span>(<span class='no'>all_markers_small</span>, <span class='kw'>n</span> <span class='kw'>=</span> <span class='fl'>1</span>)
<span class='no'>markers</span></div><div class='output co'>#&gt; # A tibble: 5 x 18
#&gt; # Groups:   cluster [5]
#&gt;   cluster rowname geneID        geneName  pct1  pct2 avgLogFC     padj   pvalue
#&gt;   &lt;fct&gt;   &lt;chr&gt;   &lt;chr&gt;         &lt;chr&gt;    &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;
#&gt; 1 0       COL1A2  ENSG00000164… COL1A2   1     0.465     2.56 5.31e-58 1.06e-60
#&gt; 2 1       S100A10 ENSG00000197… S100A10  0.741 0.266     2.17 4.75e-30 9.50e-33
#&gt; 3 2       MYL4    ENSG00000198… MYL4     0.872 0.123     2.14 3.25e-55 6.50e-58
#&gt; 4 3       MT-ATP6 ENSG00000198… MT-ATP6  1     0.993     1.16 1.80e-31 3.59e-34
#&gt; 5 4       H3F3A   ENSG00000163… H3F3A    1     0.767     1.38 3.89e-25 7.79e-28
#&gt; # ... with 9 more variables: seqnames &lt;fct&gt;, start &lt;int&gt;, end &lt;int&gt;,
#&gt; #   width &lt;int&gt;, strand &lt;fct&gt;, geneBiotype &lt;fct&gt;, description &lt;chr&gt;,
#&gt; #   seqCoordSystem &lt;fct&gt;, broadClass &lt;fct&gt;</div><div class='input'><span class='no'>markers</span> <span class='kw'>&lt;-</span> <span class='fu'>head</span>(<span class='no'>markers</span>, <span class='kw'>n</span> <span class='kw'>=</span> <span class='fl'>1</span>)
<span class='fu'>plotTopMarkers</span>(<span class='no'>object</span>, <span class='kw'>markers</span> <span class='kw'>=</span> <span class='no'>markers</span>)</div><div class='output co'>#&gt;    |                                                  | 0 % ~calculating  
#&gt; 
#&gt; ## Cluster 0 {.tabset}
#&gt; 
#&gt; 
#&gt; 
#&gt; ### COL1A2 {.tabset}
#&gt; </div><div class='img'><img src='plotMarker-6.png' alt='' width='700' height='433' /></div><div class='output co'>#&gt;    |++++++++++                                        | 20% ~04s             |++++++++++++++++++++                              | 40% ~01s             |++++++++++++++++++++++++++++++                    | 60% ~01s             |++++++++++++++++++++++++++++++++++++++++         | 80% ~00s             |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed = 01s</div><div class='input'>
<span class='co'># Known markers detected</span>
<span class='no'>markers</span> <span class='kw'>&lt;-</span> <span class='fu'>head</span>(<span class='no'>known_markers_small</span>, <span class='kw'>n</span> <span class='kw'>=</span> <span class='fl'>1</span>)
<span class='no'>markers</span></div><div class='output co'>#&gt; # A tibble: 1 x 19
#&gt; # Groups:   cellType [1]
#&gt;   cellType     cluster geneID    geneName rowname  pct1  pct2 avgLogFC     padj
#&gt;   &lt;chr&gt;        &lt;fct&gt;   &lt;chr&gt;     &lt;chr&gt;    &lt;chr&gt;   &lt;dbl&gt; &lt;dbl&gt;    &lt;dbl&gt;    &lt;dbl&gt;
#&gt; 1 Natural Kil… 1       ENSG0000… NCAM1    NCAM1   0.466 0.701   -0.954  2.25e-7
#&gt; # ... with 10 more variables: pvalue &lt;dbl&gt;, seqnames &lt;fct&gt;, start &lt;int&gt;,
#&gt; #   end &lt;int&gt;, width &lt;int&gt;, strand &lt;fct&gt;, geneBiotype &lt;fct&gt;, description &lt;chr&gt;,
#&gt; #   seqCoordSystem &lt;fct&gt;, broadClass &lt;fct&gt;</div><div class='input'><span class='fu'>plotKnownMarkersDetected</span>(<span class='no'>object</span>, <span class='kw'>markers</span> <span class='kw'>=</span> <span class='no'>markers</span>)</div><div class='output co'>#&gt;    |                                                  | 0 % ~calculating  
#&gt; 
#&gt; ## Natural Killer Cell {.tabset}
#&gt; 
#&gt; 
#&gt; 
#&gt; ### NCAM1 {.tabset}
#&gt; </div><div class='img'><img src='plotMarker-7.png' alt='' width='700' height='433' /></div><div class='output co'>#&gt;    |++++++++++++++++++++++++++++++++++++++++++++++++++| 100% elapsed = 01s</div></pre>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
    <h2>Contents</h2>
    <ul class="nav nav-pills nav-stacked">
      <li><a href="#arguments">Arguments</a></li>
      
      <li><a href="#value">Value</a></li>

      <li><a href="#plottopmarkers">plotTopMarkers</a></li>

      <li><a href="#see-also">See also</a></li>
      
      <li><a href="#examples">Examples</a></li>
    </ul>

    <h2>Author</h2>
    
Michael Steinbaugh, Rory Kirchner

  </div>
</div>

      <footer>
      <div class="copyright">
  <p>Developed by <a href='https://steinbaugh.com'>Michael Steinbaugh</a>, <a href='https://github.com/roryk'>Rory Kirchner</a>, <a href='https://github.com/marypiper'>Mary Piper</a>, <a href='https://github.com/vbarrera'>Victor Barrera</a>, <a href='https://github.com/sjhosui'>Shannan Ho Sui</a>, <a href='http://bioinformatics.sph.harvard.edu'>Harvard Chan Bioinformatics Core</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://pkgdown.r-lib.org/">pkgdown</a>.</p>
</div>

      </footer>
   </div>

  

  </body>
</html>

